// @ui5/webcomponents-react/lib/Bar.js
import _extends from '@babel/runtime/helpers/extends';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { useDeprecateRenderMethods } from '@ui5/webcomponents-react-base/lib/hooks';
import React, { forwardRef } from 'react';
import { createComponentStyles } from '@ui5/webcomponents-react-base/lib/createComponentStyles';
import { CssSizeVariables } from '@ui5/webcomponents-react-base/lib/CssSizeVariables';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';

var BarDesign;

(function (BarDesign) {
  BarDesign["Auto"] = "Auto";
  BarDesign["Footer"] = "Footer";
  BarDesign["Header"] = "Header";
  BarDesign["SubHeader"] = "SubHeader";
  BarDesign["FloatingFooter"] = "FloatingFooter";
})(BarDesign || (BarDesign = {}));

var styles = {
  // outer container, controlling height and width
  bar: {
    width: '100%',
    display: 'block',
    position: 'relative',
    height: CssSizeVariables.sapWcrBarHeight
  },
  // left container
  left: {
    position: 'absolute',
    left: '0',
    top: '0',
    height: '100%',
    paddingLeft: '0.5rem',
    textAlign: 'left',
    display: 'flex',
    alignItems: 'center'
  },
  // center container
  center: {
    textAlign: 'center',
    width: '100%',
    height: '0',
    top: '0',
    left: '0'
  },
  // inner container around left, center, and right
  inner: {
    display: 'inline-flex',
    padding: '0 0.5rem 0 0.5rem',
    alignItems: 'center'
  },
  // right container
  right: {
    position: 'absolute',
    right: '0',
    top: '0',
    height: '100%',
    paddingRight: '0.5rem',
    textAlign: 'right',
    display: 'flex',
    alignItems: 'center'
  },
  auto: {
    height: CssSizeVariables.sapWcrBarHeight,
    background: ThemingParameters.sapPageHeader_Background,
    boxShadow: ThemingParameters.sapContent_HeaderShadow,
    '& $inner': {
      height: CssSizeVariables.sapWcrBarHeight
    }
  },
  subHeader: {
    height: CssSizeVariables.sapWcrSubHeaderBarHeight,
    background: ThemingParameters.sapPageHeader_Background,
    boxShadow: ThemingParameters.sapContent_HeaderShadow,
    paddingBottom: ' 0.25rem',
    '& $inner': {
      height: CssSizeVariables.sapWcrSubHeaderBarHeight
    }
  },
  footer: {
    height: CssSizeVariables.sapWcrBarHeight,
    background: ThemingParameters.sapPageFooter_Background,
    borderTop: "0.0625rem solid ".concat(ThemingParameters.sapPageFooter_BorderColor),
    '& $inner': {
      height: CssSizeVariables.sapWcrBarHeight
    }
  },
  floatingFooter: {
    height: CssSizeVariables.sapWcrBarHeight,
    background: ThemingParameters.sapPageFooter_Background,
    // borderRadius: ThemingParameters.sapElement_BorderCornerRadius,
    boxShadow: ThemingParameters.sapContent_Shadow1,
    border: 'none',
    opacity: 1,
    '& $inner': {
      height: CssSizeVariables.sapWcrBarHeight
    }
  }
};

var useStyles = createComponentStyles(styles, {
  name: 'Bar'
});
/**
 * <code>import { Bar } from '@ui5/webcomponents-react/lib/Bar';</code>
 */

var Bar = forwardRef(function (props, ref) {
  var className = props.className,
      style = props.style,
      tooltip = props.tooltip,
      slot = props.slot,
      design = props.design;
  var contentLeft = useDeprecateRenderMethods(props, 'renderContentLeft', 'contentLeft');
  var contentMiddle = useDeprecateRenderMethods(props, 'renderContentMiddle', 'contentMiddle');
  var contentRight = useDeprecateRenderMethods(props, 'renderContentRight', 'contentRight');
  var classes = useStyles();
  var cssClasses = StyleClassHelper.of(classes.bar);

  switch (design) {
    case BarDesign.Footer:
      cssClasses.put(classes.footer);
      break;

    case BarDesign.SubHeader:
      cssClasses.put(classes.subHeader);
      break;

    case BarDesign.FloatingFooter:
      cssClasses.put(classes.floatingFooter);
      break;

    case BarDesign.Header:
    case BarDesign.Auto:
    default:
      cssClasses.put(classes.auto);
  }

  if (className) {
    cssClasses.put(className);
  }

  var passThroughProps = usePassThroughHtmlProps(props);
  return /*#__PURE__*/React.createElement("div", _extends({
    "data-bar-part": "Root",
    className: cssClasses.toString(),
    style: style,
    title: tooltip,
    slot: slot,
    ref: ref
  }, passThroughProps), /*#__PURE__*/React.createElement("div", {
    "data-bar-part": "Left",
    className: classes.left
  }, contentLeft), /*#__PURE__*/React.createElement("div", {
    "data-bar-part": "Center",
    className: classes.center
  }, /*#__PURE__*/React.createElement("div", {
    className: classes.inner
  }, contentMiddle)), /*#__PURE__*/React.createElement("div", {
    "data-bar-part": "Right",
    className: classes.right
  }, contentRight));
});
Bar.displayName = 'Bar';
Bar.defaultProps = {
  design: BarDesign.Auto
};

export { Bar };
