// @ui5/webcomponents-react/lib/DateTimePicker.js
import { CalendarType } from '@ui5/webcomponents-react/lib/CalendarType';
import { ValueState } from '@ui5/webcomponents-react/lib/ValueState';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/DateTimePicker';

/**
 * <code>import { DateTimePicker } from '@ui5/webcomponents-react/lib/DateTimePicker';</code>
 * <br />
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/DateTimePicker" target="_blank">UI5 Web Components Playground</a>
 */
var DateTimePicker = withWebComponent('ui5-datetime-picker', ['formatPattern', 'maxDate', 'minDate', 'name', 'placeholder', 'primaryCalendarType', 'value', 'valueState'], ['disabled', 'readonly'], [], ['change', 'input']);
DateTimePicker.displayName = 'DateTimePicker';
DateTimePicker.defaultProps = {
  disabled: false,
  placeholder: undefined,
  primaryCalendarType: CalendarType.Gregorian,
  readonly: false,
  valueState: ValueState.None
};

export { DateTimePicker };
