// @ui5/webcomponents-react/lib/FormGroup.js
import { createComponentStyles } from '@ui5/webcomponents-react-base/lib/createComponentStyles';
import { Title } from '@ui5/webcomponents-react/lib/Title';
import { TitleLevel } from '@ui5/webcomponents-react/lib/TitleLevel';
import React, { forwardRef, Children } from 'react';

var useStyles = createComponentStyles({
  title: {
    paddingBottom: '1em'
  },
  formGroup: {
    width: '100%',
    paddingTop: '0.25em',
    display: 'block'
  }
}, {
  name: 'FormGroup'
});
/**
 * <code>import { FormGroup } from '@ui5/webcomponents-react/lib/FormGroup';</code>
 */

var FormGroup = forwardRef(function (props, ref) {
  var title = props.title,
      children = props.children,
      tooltip = props.tooltip,
      style = props.style,
      className = props.className,
      slot = props.slot;
  var classes = useStyles();
  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    style: style,
    className: className,
    title: tooltip,
    slot: slot
  }, title && /*#__PURE__*/React.createElement(Title, {
    level: TitleLevel.H5,
    className: classes.title
  }, title), Children.map(children, function (child, index) {
    return /*#__PURE__*/React.createElement("div", {
      key: index,
      className: classes.formGroup
    }, child);
  }));
});
FormGroup.displayName = 'FormGroup';

export { FormGroup };
