// @ui5/webcomponents-react/lib/MessageBox.js
import _extends from '@babel/runtime/helpers/extends';
import '@ui5/webcomponents-icons/dist/icons/hint';
import '@ui5/webcomponents-icons/dist/icons/message-error';
import '@ui5/webcomponents-icons/dist/icons/message-information';
import '@ui5/webcomponents-icons/dist/icons/message-success';
import '@ui5/webcomponents-icons/dist/icons/message-warning';
import '@ui5/webcomponents-icons/dist/icons/question-mark';
import { createComponentStyles } from '@ui5/webcomponents-react-base/lib/createComponentStyles';
import { useI18nBundle, useConsolidatedRef, usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/hooks';
import { enrichEventWithDetails } from '@ui5/webcomponents-react-base/lib/Utils';
import { ABORT, CANCEL, CLOSE, DELETE, IGNORE, NO, OK, RETRY, YES, HIGHLIGHT, WARNING, SUCCESS, INFORMATION, ERROR, CONFIRMATION } from '@ui5/webcomponents-react/dist/assets/i18n/i18n-defaults';
import { Button } from '@ui5/webcomponents-react/lib/Button';
import { ButtonDesign } from '@ui5/webcomponents-react/lib/ButtonDesign';
import { Dialog } from '@ui5/webcomponents-react/lib/Dialog';
import { Icon } from '@ui5/webcomponents-react/lib/Icon';
import { MessageBoxActions } from '@ui5/webcomponents-react/lib/MessageBoxActions';
import { MessageBoxTypes } from '@ui5/webcomponents-react/lib/MessageBoxTypes';
import { Text } from '@ui5/webcomponents-react/lib/Text';
import { Title } from '@ui5/webcomponents-react/lib/Title';
import { TitleLevel } from '@ui5/webcomponents-react/lib/TitleLevel';
import React, { forwardRef, useMemo, isValidElement, useCallback, useEffect } from 'react';
import { CssSizeVariables } from '@ui5/webcomponents-react-base/lib/CssSizeVariables';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';

var style = {
  header: {
    display: 'flex',
    height: CssSizeVariables.sapWcrBarHeight,
    lineHeight: CssSizeVariables.sapWcrBarHeight,
    textAlign: 'center',
    width: 'auto',
    alignItems: 'center',
    '& $icon': {
      '& >*': {
        verticalAlign: 'middle'
      }
    },
    // justifyContent: 'center',
    padding: '0.25rem 1rem',
    boxSizing: 'border-box',
    borderBottom: "1px solid var(--messageBoxBorderColor)",
    color: ThemingParameters.sapContent_LabelColor,
    fontSize: '1rem',
    '&[data-type="Error"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapErrorBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapErrorBorderColor,
      '& $icon': {
        '--sapContent_NonInteractiveIconColor': ThemingParameters.sapNegativeElementColor
      }
    },
    '&[data-type="Warning"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapWarningBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapWarningBorderColor,
      '& $icon': {
        '--sapContent_NonInteractiveIconColor': ThemingParameters.sapCriticalElementColor
      }
    },
    '&[data-type="Success"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapSuccessBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapSuccessBorderColor,
      '& $icon': {
        '--sapContent_NonInteractiveIconColor': ThemingParameters.sapPositiveElementColor
      }
    },
    '&[data-type="Confirm"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapNeutralBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapNeutralBorderColor,
      '& $icon': {
        '--sapContent_NonInteractiveIconColor': ThemingParameters.sapNeutralElementColor
      }
    },
    '&[data-type="Information"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapNeutralBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapNeutralBorderColor,
      '& $icon': {
        '--sapContent_NonInteractiveIconColor': ThemingParameters.sapNeutralElementColor
      }
    },
    '&[data-type="Highlight"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapInformationBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapInformationBorderColor,
      '& $icon': {
        '--sapContent_NonInteractiveIconColor': ThemingParameters.sapInformativeElementColor
      }
    }
  },
  icon: {
    marginRight: '0.5rem',
    fontSize: '1rem',
    width: '1rem'
  },
  content: {
    padding: '1rem'
  },
  footer: {
    height: CssSizeVariables.sapWcrBarHeight,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
    padding: '0 0.5rem',
    '& >*:not(:last-child)': {
      marginRight: '0.5rem'
    },
    '& > ui5-button': {
      display: 'flex'
    }
  }
};

var actionTextMap = new Map();
actionTextMap.set(MessageBoxActions.ABORT, ABORT);
actionTextMap.set(MessageBoxActions.CANCEL, CANCEL);
actionTextMap.set(MessageBoxActions.CLOSE, CLOSE);
actionTextMap.set(MessageBoxActions.DELETE, DELETE);
actionTextMap.set(MessageBoxActions.IGNORE, IGNORE);
actionTextMap.set(MessageBoxActions.NO, NO);
actionTextMap.set(MessageBoxActions.OK, OK);
actionTextMap.set(MessageBoxActions.RETRY, RETRY);
actionTextMap.set(MessageBoxActions.YES, YES);
var useStyles = createComponentStyles(style, {
  name: 'MessageBox'
});
/**
 * <code>import { MessageBox } from '@ui5/webcomponents-react/lib/MessageBox';</code>
 */

var MessageBox = forwardRef(function (props, ref) {
  var open = props.open,
      type = props.type,
      children = props.children,
      className = props.className,
      style = props.style,
      tooltip = props.tooltip,
      slot = props.slot,
      title = props.title,
      icon = props.icon,
      actions = props.actions,
      onClose = props.onClose;
  var i18nBundle = useI18nBundle('@ui5/webcomponents-react');
  var classes = useStyles();
  var iconToRender = useMemo(function () {
    if (isValidElement(icon)) return icon;

    switch (type) {
      case MessageBoxTypes.CONFIRM:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "question-mark"
        });

      case MessageBoxTypes.ERROR:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "message-error"
        });

      case MessageBoxTypes.INFORMATION:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "message-information"
        });

      case MessageBoxTypes.SUCCESS:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "message-success"
        });

      case MessageBoxTypes.WARNING:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "message-warning"
        });

      case MessageBoxTypes.HIGHLIGHT:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "hint"
        });
    }

    return null;
  }, [icon, type]);
  var titleToRender = useMemo(function () {
    if (title) {
      return title;
    }

    switch (type) {
      case MessageBoxTypes.CONFIRM:
        return i18nBundle.getText(CONFIRMATION);

      case MessageBoxTypes.ERROR:
        return i18nBundle.getText(ERROR);

      case MessageBoxTypes.INFORMATION:
        return i18nBundle.getText(INFORMATION);

      case MessageBoxTypes.SUCCESS:
        return i18nBundle.getText(SUCCESS);

      case MessageBoxTypes.WARNING:
        return i18nBundle.getText(WARNING);

      case MessageBoxTypes.HIGHLIGHT:
        return i18nBundle.getText(HIGHLIGHT);
    }

    return null;
  }, [title, type, i18nBundle]);
  var actionsToRender = useMemo(function () {
    if (actions && actions.length > 0) {
      return actions;
    }

    if (type === MessageBoxTypes.CONFIRM) {
      return [MessageBoxActions.OK, MessageBoxActions.CANCEL];
    }

    if (type === MessageBoxTypes.ERROR) {
      return [MessageBoxActions.CLOSE];
    }

    return [MessageBoxActions.OK];
  }, [actions, type]);
  var handleOnClose = useCallback(function (e) {
    var action = e.target.dataset.action;
    onClose(enrichEventWithDetails(e, {
      action: action
    }));
  }, [onClose]);
  var dialogRef = useConsolidatedRef(ref);
  useEffect(function () {
    if (dialogRef.current) {
      if (open) {
        dialogRef.current.open();
      } else {
        dialogRef.current.close();
      }
    }
  }, [open, dialogRef]);
  var passThroughProps = usePassThroughHtmlProps(props, ['onClose']);
  return /*#__PURE__*/React.createElement(Dialog, _extends({
    slot: slot,
    ref: dialogRef,
    style: style,
    tooltip: tooltip,
    className: className,
    onAfterClose: open ? handleOnClose : null,
    header: /*#__PURE__*/React.createElement("header", {
      className: classes.header,
      "data-type": type
    }, !!iconToRender && /*#__PURE__*/React.createElement("div", {
      className: classes.icon
    }, iconToRender), /*#__PURE__*/React.createElement(Title, {
      level: TitleLevel.H5
    }, titleToRender)),
    footer: /*#__PURE__*/React.createElement("footer", {
      className: classes.footer
    }, actionsToRender.map(function (action, index) {
      var text = i18nBundle.getText(actionTextMap.get(action));
      return /*#__PURE__*/React.createElement(Button, {
        style: {
          minWidth: '4rem'
        },
        key: action,
        design: index === 0 ? ButtonDesign.Emphasized : ButtonDesign.Transparent,
        onClick: handleOnClose,
        "data-action": action
      }, text);
    }))
  }, passThroughProps), /*#__PURE__*/React.createElement(Text, {
    className: classes.content
  }, children));
});
MessageBox.displayName = 'MessageBox';
MessageBox.defaultProps = {
  open: false,
  title: null,
  icon: null,
  type: MessageBoxTypes.CONFIRM,
  actions: []
};

export { MessageBox };
