// @ui5/webcomponents-react/lib/Page.js
import _extends from '@babel/runtime/helpers/extends';
import '@ui5/webcomponents-icons/dist/icons/navigation-left-arrow';
import { enrichEventWithDetails } from '@ui5/webcomponents-react-base/lib/Utils';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { useDeprecateRenderMethods } from '@ui5/webcomponents-react-base/lib/hooks';
import { Bar } from '@ui5/webcomponents-react/lib/Bar';
import { Button } from '@ui5/webcomponents-react/lib/Button';
import { ButtonDesign } from '@ui5/webcomponents-react/lib/ButtonDesign';
import { PageBackgroundDesign } from '@ui5/webcomponents-react/lib/PageBackgroundDesign';
import { Title } from '@ui5/webcomponents-react/lib/Title';
import { TitleLevel } from '@ui5/webcomponents-react/lib/TitleLevel';
import React, { forwardRef, useCallback, useMemo } from 'react';
import { createComponentStyles } from '@ui5/webcomponents-react-base/lib/createComponentStyles';
import { CssSizeVariables } from '@ui5/webcomponents-react-base/lib/CssSizeVariables';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';

/**
 * Style Class Generator Function
 *
 * @param {Object} obj - Current Theme Context by JSS Provider.
 * @param {object} obj.parameters - Theming parameters (e.g. LabelColor)
 */

var styles = {
  // const isS = '@media (max-width: 600px)';
  // const isM = '@media (min-width: 601px) and (max-width: 1024px)';
  // const isL = '@media (min-width: 1025px) and (max-width: 1440px)';
  // const isXL = '@media (min-width: 1441px) and (max-width: 1920px)';
  // const isXXL = '(min-width: 1921px)';
  // function getHeaderAndFooterPaddings() {
  //   return {
  //     '& [data-bar-part="Left"]': {
  //       [isS]: {
  //         paddingLeft: '1rem'
  //       },
  //       [isM]: {
  //         paddingLeft: '2rem'
  //       },
  //       [isL]: {
  //         paddingLeft: '2rem'
  //       },
  //       [isXL]: {
  //         paddingLeft: '3rem'
  //       },
  //       [isXXL]: {
  //         paddingLeft: '3rem'
  //       }
  //     },
  //     '& [data-bar-part="Right"]': {
  //       [isS]: {
  //         paddingRight: '1rem'
  //       },
  //       [isM]: {
  //         paddingRight: '2rem'
  //       },
  //       [isL]: {
  //         paddingRight: '2rem'
  //       },
  //       [isXL]: {
  //         paddingRight: '3rem'
  //       },
  //       [isXXL]: {
  //         paddingRight: '3rem'
  //       }
  //     }
  //   };
  // }
  pageContainer: {
    width: '100%',
    height: '100%',
    overflow: 'hidden',
    position: 'relative',
    '& $pageHeader': {// ...getHeaderAndFooterPaddings()
    },
    '& $pageFooter': {// ...getHeaderAndFooterPaddings()
    },
    '& $contentSection': {// [isS]: {
      //   padding: '1rem 1rem 0 1rem'
      // },
      // [isM]: {
      //   padding: '1rem 2rem 0 2rem'
      // },
      // [isL]: {
      //   padding: '1rem 2rem 0 2rem'
      // },
      // [isXL]: {
      //   padding: '1rem 3rem 0 3rem'
      // },
      // [isXXL]: {
      //   padding: '1rem 3rem 0 3rem'
      // }
    }
  },
  backgroundStandard: {
    background: ThemingParameters.sapBackgroundColor
  },
  backgroundSolid: {
    background: ThemingParameters.sapShell_Background
  },
  backgroundList: {
    background: ThemingParameters.sapGroup_ContentBackground
  },
  backgroundTransparent: {
    background: 'transparent'
  },
  baseBar: {
    '& >*': {
      paddingLeft: '0rem',
      paddingRight: '0rem',
      '& [data-bar-part="Left"]': {
        paddingLeft: '0rem'
      },
      '& [data-bar-part="Right"]': {
        paddingRight: '0rem'
      }
    }
  },
  pageHeader: {
    top: 0,
    width: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    textAlign: 'center',
    background: ThemingParameters.sapPageHeader_Background,
    '& >*': {
      background: ThemingParameters.sapPageHeader_Background,
      boxShadow: "inset 0 -0.0625rem ".concat(ThemingParameters.sapPageHeader_BorderColor)
    }
  },
  contentSection: {
    position: 'absolute',
    top: '0',
    left: '0',
    right: '0',
    bottom: '0',
    overflowY: 'auto',
    overflowX: 'hidden',
    // marginTop: '1px',
    // marginBottom: '1px',
    boxSizing: 'border-box' // backgroundColor: ThemingParameters.sapBackgroundColor

  },
  pageWithHeader: {
    '& $contentSection': {
      top: CssSizeVariables.sapWcrBarHeight
    }
  },
  pageWithFooter: {
    '& $contentSection': {
      bottom: CssSizeVariables.sapWcrBarHeight
    }
  },
  pageFooter: {
    position: 'absolute',
    bottom: '0',
    left: '0',
    width: '100%',
    borderTop: "1px solid ".concat(ThemingParameters.sapPageFooter_BorderColor),
    '& >*': {
      height: CssSizeVariables.sapWcrBarHeight,
      background: ThemingParameters.sapPageFooter_Background,
      borderTop: "0.0625rem solid ".concat(ThemingParameters.sapPageFooter_BorderColor)
    }
  }
};

var useStyles = createComponentStyles(styles, {
  name: 'Page'
});
/**
 * <code>import { Page } from '@ui5/webcomponents-react/lib/Page';</code>
 */

var Page = forwardRef(function (props, ref) {
  var children = props.children,
      showFooter = props.showFooter,
      showHeader = props.showHeader,
      showBackButton = props.showBackButton,
      className = props.className,
      style = props.style,
      backgroundDesign = props.backgroundDesign,
      tooltip = props.tooltip,
      slot = props.slot,
      onNavButtonPress = props.onNavButtonPress,
      title = props.title;
  var customHeader = useDeprecateRenderMethods(props, 'renderCustomHeader', 'customHeader');
  var customFooter = useDeprecateRenderMethods(props, 'renderCustomFooter', 'customFooter');
  var classes = useStyles();
  var handleNavBackButtonPress = useCallback(function (e) {
    if (typeof onNavButtonPress === 'function') {
      onNavButtonPress(enrichEventWithDetails(e));
    }
  }, [onNavButtonPress]);
  var renderBackButton = useCallback(function () {
    if (showBackButton) {
      return /*#__PURE__*/React.createElement(Button, {
        icon: "navigation-left-arrow",
        design: ButtonDesign.Transparent,
        onClick: handleNavBackButtonPress
      });
    }

    return null;
  }, [showBackButton]);
  var renderTitle = useCallback(function () {
    return /*#__PURE__*/React.createElement(Title, {
      level: TitleLevel.H5
    }, title);
  }, [title]);
  var header = useMemo(function () {
    return customHeader !== null && customHeader !== void 0 ? customHeader : /*#__PURE__*/React.createElement(Bar, {
      contentLeft: renderBackButton(),
      contentMiddle: renderTitle()
    });
  }, [customHeader, renderTitle, renderBackButton]);
  var pageContainer = StyleClassHelper.of(classes.pageContainer);
  var headerClasses = StyleClassHelper.of(classes.pageHeader, classes.baseBar);
  var footerClasses = StyleClassHelper.of(classes.pageFooter, classes.baseBar);

  if (showHeader) {
    pageContainer.put(classes.pageWithHeader);
  }

  if (showFooter) {
    pageContainer.put(classes.pageWithFooter);
  }

  if (className) {
    pageContainer.put(className);
  }

  pageContainer.put(classes["background".concat(backgroundDesign)]);
  var passThroughProps = usePassThroughHtmlProps(props, ['onNavButtonPress']);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: pageContainer.valueOf(),
    style: style,
    title: tooltip,
    slot: slot
  }, passThroughProps), showHeader && /*#__PURE__*/React.createElement("header", {
    className: headerClasses.valueOf()
  }, header), /*#__PURE__*/React.createElement("section", {
    className: classes.contentSection
  }, children), showFooter && /*#__PURE__*/React.createElement("footer", {
    className: footerClasses.valueOf()
  }, customFooter));
});
Page.defaultProps = {
  showHeader: true,
  showFooter: false,
  showBackButton: true,
  renderCustomHeader: null,
  renderCustomFooter: null,
  title: '',
  backgroundDesign: PageBackgroundDesign.Standard
};
Page.displayName = 'Page';

export { Page };
