// @ui5/webcomponents-react/lib/Popover.js
import { PlacementType } from '@ui5/webcomponents-react/lib/PlacementType';
import { PopoverHorizontalAlign } from '@ui5/webcomponents-react/lib/PopoverHorizontalAlign';
import { PopoverVerticalAlign } from '@ui5/webcomponents-react/lib/PopoverVerticalAlign';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/Popover';

/**
 * <code>import { Popover } from '@ui5/webcomponents-react/lib/Popover';</code>
 * <br />
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/Popover" target="_blank">UI5 Web Components Playground</a>
 */
var Popover = withWebComponent('ui5-popover', ['headerText', 'horizontalAlign', 'initialFocus', 'placementType', 'verticalAlign'], ['allowTargetOverlap', 'modal', 'noArrow'], ['footer', 'header'], ['afterClose', 'afterOpen', 'beforeClose', 'beforeOpen']);
Popover.displayName = 'Popover';
Popover.defaultProps = {
  allowTargetOverlap: false,
  horizontalAlign: PopoverHorizontalAlign.Center,
  modal: false,
  noArrow: false,
  placementType: PlacementType.Right,
  verticalAlign: PopoverVerticalAlign.Center
};

export { Popover };
