// @ui5/webcomponents-react/lib/ResponsivePopover.js
import { PlacementType } from '@ui5/webcomponents-react/lib/PlacementType';
import { PopoverHorizontalAlign } from '@ui5/webcomponents-react/lib/PopoverHorizontalAlign';
import { PopoverVerticalAlign } from '@ui5/webcomponents-react/lib/PopoverVerticalAlign';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/ResponsivePopover';

/**
 * <code>import { ResponsivePopover } from '@ui5/webcomponents-react/lib/ResponsivePopover';</code>
 * <br />
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/ResponsivePopover" target="_blank">UI5 Web Components Playground</a>
 */
var ResponsivePopover = withWebComponent('ui5-responsive-popover', ['headerText', 'horizontalAlign', 'initialFocus', 'placementType', 'verticalAlign'], ['allowTargetOverlap', 'modal', 'noArrow'], ['footer', 'header'], ['afterClose', 'afterOpen', 'beforeClose', 'beforeOpen']);
ResponsivePopover.displayName = 'ResponsivePopover';
ResponsivePopover.defaultProps = {
  allowTargetOverlap: false,
  horizontalAlign: PopoverHorizontalAlign.Center,
  modal: false,
  noArrow: false,
  placementType: PlacementType.Right,
  verticalAlign: PopoverVerticalAlign.Center
};

export { ResponsivePopover };
