// @ui5/webcomponents-react/lib/UploadCollectionItem.js
import '@ui5/webcomponents-fiori/dist/UploadCollectionItem';
import { UploadState } from '@ui5/webcomponents-react/lib/UploadState';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';

/**
 * <code>import { UploadCollectionItem } from '@ui5/webcomponents-react/lib/UploadCollectionItem';</code>
 * <br />
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/UploadCollectionItem" target="_blank">UI5 Web Components Playground</a>
 */
var UploadCollectionItem = withWebComponent('ui5-upload-collection-item', ['file', 'fileName', 'progress', 'uploadState'], ['fileNameClickable', 'noDelete', 'noRetry', 'noTerminate'], ['thumbnail'], ['fileNameClick', 'rename', 'retry', 'terminate']);
UploadCollectionItem.displayName = 'UploadCollectionItem';
UploadCollectionItem.defaultProps = {
  file: null,
  fileNameClickable: false,
  noDelete: false,
  noRetry: false,
  noTerminate: false,
  progress: 0,
  uploadState: UploadState.Ready
};

export { UploadCollectionItem };
