// @ui5/webcomponents-react-charts/lib/BarChartPlaceholder.js
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';
import React from 'react';
import ContentLoader from 'react-content-loader';

var BarChartPlaceholder = function BarChartPlaceholder() {
  return /*#__PURE__*/React.createElement(ContentLoader, {
    viewBox: "0 0 165 145",
    preserveAspectRatio: "xMidYMid meet",
    width: "100%",
    height: "100%",
    speed: 2,
    backgroundColor: ThemingParameters.sapContent_ImagePlaceholderBackground,
    foregroundColor: ThemingParameters.sapContent_ImagePlaceholderForegroundColor,
    backgroundOpacity: ThemingParameters.sapContent_DisabledOpacity
  }, /*#__PURE__*/React.createElement("rect", {
    x: "20",
    y: "10",
    width: "1",
    height: "135"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "20",
    y: "20",
    width: "85",
    height: "15"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "20",
    y: "40",
    width: "117",
    height: "15"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "20",
    y: "60",
    width: "67",
    height: "15"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "20",
    y: "80",
    width: "75",
    height: "15"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "20",
    y: "100",
    width: "125",
    height: "15"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "20",
    y: "120",
    width: "60",
    height: "15"
  }));
};

export { BarChartPlaceholder };
