// @ui5/webcomponents-react-charts/lib/ColumnChartPlaceholder.js
import _extends from '@babel/runtime/helpers/extends';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';
import React from 'react';
import ContentLoader from 'react-content-loader';

var ColumnChartPlaceholder = function ColumnChartPlaceholder(props) {
  return /*#__PURE__*/React.createElement(ContentLoader, _extends({
    viewBox: "0 0 165 145",
    preserveAspectRatio: "xMidYMid meet",
    width: "100%",
    height: "100%",
    speed: 2,
    backgroundColor: ThemingParameters.sapContent_ImagePlaceholderBackground,
    foregroundColor: ThemingParameters.sapContent_ImagePlaceholderForegroundColor,
    backgroundOpacity: ThemingParameters.sapContent_DisabledOpacity
  }, props), /*#__PURE__*/React.createElement("rect", {
    x: "10",
    y: "135",
    width: "135",
    height: "1"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "20",
    y: "50",
    width: "15",
    height: "85"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "40",
    y: "18",
    width: "15",
    height: "117"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "60",
    y: "68",
    width: "15",
    height: "67"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "80",
    y: "60",
    width: "15",
    height: "75"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "100",
    y: "10",
    width: "15",
    height: "125"
  }), /*#__PURE__*/React.createElement("rect", {
    x: "120",
    y: "75",
    width: "15",
    height: "60"
  }));
};

export { ColumnChartPlaceholder };
