// @ui5/webcomponents-react-charts/lib/RadarChart.js
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import { deprecationNotice } from '@ui5/webcomponents-react-base/lib/Utils';
import { useConsolidatedRef } from '@ui5/webcomponents-react-base/lib/useConsolidatedRef';
import { withChartContainer } from '@ui5/webcomponents-react-charts/lib/withChartContainer';
import React, { useCallback, useEffect, useMemo, forwardRef } from 'react';
import { Radar } from 'react-chartjs-2';
import { getTheme } from '@ui5/webcomponents-base/dist/config/Theme';
import '@ui5/webcomponents/dist/Label';
import merge from 'lodash.merge';

/* eslint-disable no-underscore-dangle */
var useLegend = function useLegend(chartRef, legendRef, noLegend, handleLegendItemPress) {
  useEffect(function () {
    if (chartRef && legendRef) {
      if (noLegend) {
        legendRef.current.innerHTML = '';
      } else {
        legendRef.current.innerHTML = chartRef.current.chartInstance.generateLegend();
        legendRef.current.querySelectorAll('li').forEach(function (legendItem) {
          legendItem.addEventListener('click', handleLegendItemPress);
        });
      }
    }
  }, [chartRef.current, legendRef.current, noLegend, handleLegendItemPress]);
};
var useLegendItemClickHandler = function useLegendItemClickHandler(chartRef, legendRef) {
  return useCallback(function (e) {
    var clickTarget = e.currentTarget;
    var datasetIndex = parseInt(clickTarget.dataset.datasetindex);
    var chartInstance = chartRef.current.chartInstance;
    var meta = chartInstance.getDatasetMeta(datasetIndex);
    meta.hidden = meta.hidden === null ? !chartInstance.data.datasets[datasetIndex].hidden : null;
    chartInstance.update();
    clickTarget.style.textDecoration = meta.hidden ? 'line-through' : 'unset';
  }, [legendRef.current, chartRef.current]);
};

var ChartBaseDefaultProps = {
  labels: [],
  datasets: [],
  colors: [],
  height: 300,
  minHeight: 300,
  minWidth: 300,
  width: 300,
  options: {},
  categoryAxisFormatter: function categoryAxisFormatter(d) {
    return d;
  },
  valueAxisFormatter: function valueAxisFormatter(d) {
    return d;
  },
  getDatasetAtEvent: function getDatasetAtEvent(d, e) {// do nothing
  },
  getElementAtEvent: function getElementAtEvent(d, e) {// do nothing
  }
};

var sequentialColors = {
  sapUiChartAccent1: '#5899DA',
  sapUiChartAccent2: '#E8743B',
  sapUiChartAccent3: '#19A979',
  sapUiChartAccent4: '#ED4A7B',
  sapUiChartAccent5: '#945ECF',
  sapUiChartAccent6: '#13A4B4',
  sapUiChartAccent7: '#525DF4',
  sapUiChartAccent8: '#BF399E',
  sapUiChartAccent9: '#6C8893',
  sapUiChartAccent10: '#EE6868',
  sapUiChartAccent11: '#2F6497',
  sapUiChartAccent12: '#1866b4'
};
var semanticColors = {
  sapUiChartGood: '#3fa45b',
  sapUiChartBad: '#dc0d0e',
  sapUiChartHighlight: '#de890d'
};

var sapBelize = /*#__PURE__*/Object.freeze({
  __proto__: null,
  sequentialColors: sequentialColors,
  semanticColors: semanticColors
});

var sequentialColors$1 = {
  sapUiChartAccent1: '#5899DA',
  sapUiChartAccent2: '#E8743B',
  sapUiChartAccent3: '#19A979',
  sapUiChartAccent4: '#ED4A7B',
  sapUiChartAccent5: '#945ECF',
  sapUiChartAccent6: '#13A4B4',
  sapUiChartAccent7: '#525DF4',
  sapUiChartAccent8: '#BF399E',
  sapUiChartAccent9: '#6C8893',
  sapUiChartAccent10: '#EE6868',
  sapUiChartAccent11: '#2F6497',
  sapUiChartAccent12: '#1866b4'
};
var semanticColors$1 = {
  sapUiChartGood: '#69c683',
  sapUiChartBad: '#f55556',
  sapUiChartHighlight: '#f4ac44'
};

var sapBelizePlus = /*#__PURE__*/Object.freeze({
  __proto__: null,
  sequentialColors: sequentialColors$1,
  semanticColors: semanticColors$1
});

var sequentialColors$2 = {
  sapUiChartAccent1: '#0f828f',
  sapUiChartAccent2: '#5ac2ce',
  sapUiChartAccent3: '#03734d',
  sapUiChartAccent4: '#66c2a3',
  sapUiChartAccent5: '#3c6372',
  sapUiChartAccent6: '#adbcc3',
  sapUiChartAccent7: '#144b7f',
  sapUiChartAccent8: '#698caf',
  sapUiChartAccent9: '#d62f2f',
  sapUiChartAccent10: '#f8a6a6',
  sapUiChartAccent11: '#921473',
  sapUiChartAccent12: '#d876c0'
};
var semanticColors$2 = {
  sapUiChartGood: '#3fa45b',
  sapUiChartBad: '#dc0d0e',
  sapUiChartHighlight: '#f0ab00'
};

var sapFiori3 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  sequentialColors: sequentialColors$2,
  semanticColors: semanticColors$2
});

var sequentialColors$3 = {
  sapUiChartAccent1: '#13a4b4',
  sapUiChartAccent2: '#81d1da',
  sapUiChartAccent3: '#19a979',
  sapUiChartAccent4: '#8fd1bb',
  sapUiChartAccent5: '#6c8893',
  sapUiChartAccent6: '#d1d9dc',
  sapUiChartAccent7: '#2f6497',
  sapUiChartAccent8: '#85a1bb',
  sapUiChartAccent9: '#ee6868',
  sapUiChartAccent10: '#fccaca',
  sapUiChartAccent11: '#bf399e',
  sapUiChartAccent12: '#e597d2'
};
var semanticColors$3 = {
  sapUiChartGood: '#3fa45b',
  sapUiChartBad: '#dc0d0e',
  sapUiChartHighlight: '#f0ab00'
};

var sapFiori3Dark = /*#__PURE__*/Object.freeze({
  __proto__: null,
  sequentialColors: sequentialColors$3,
  semanticColors: semanticColors$3
});

var sequentialColors$4 = {
  sapUiChartAccent1: '#4DA2F6',
  sapUiChartAccent2: '#E8743B',
  sapUiChartAccent3: '#19A979',
  sapUiChartAccent4: '#ED4A7B',
  sapUiChartAccent5: '#945ECF',
  sapUiChartAccent6: '#13A4B4',
  sapUiChartAccent7: '#5C66F5',
  sapUiChartAccent8: '#C63FA4',
  sapUiChartAccent9: '#6C8893',
  sapUiChartAccent10: '#EE6868',
  sapUiChartAccent11: '#387BB7',
  sapUiChartAccent12: '#097be8'
};
var semanticColors$4 = {
  sapUiChartGood: '#9c9',
  sapUiChartBad: '#ff5e5e',
  sapUiChartHighlight: '#ffab1d'
};

var sapBelizeHCB = /*#__PURE__*/Object.freeze({
  __proto__: null,
  sequentialColors: sequentialColors$4,
  semanticColors: semanticColors$4
});

var sequentialColors$5 = {
  sapUiChartAccent1: '#286AC3',
  sapUiChartAccent2: '#D74114',
  sapUiChartAccent3: '#057D47',
  sapUiChartAccent4: '#E11E48',
  sapUiChartAccent5: '#642EB2',
  sapUiChartAccent6: '#03778C',
  sapUiChartAccent7: '#232DEB',
  sapUiChartAccent8: '#9B136F',
  sapUiChartAccent9: '#395563',
  sapUiChartAccent10: '#E02E2E',
  sapUiChartAccent11: '#0D3367',
  sapUiChartAccent12: '#0e4186'
};
var semanticColors$5 = {
  sapUiChartGood: '#006362',
  sapUiChartBad: '#ab0000',
  sapUiChartHighlight: '#5c5c00'
};

var sapBelizeHCW = /*#__PURE__*/Object.freeze({
  __proto__: null,
  sequentialColors: sequentialColors$5,
  semanticColors: semanticColors$5
});

var themeMap = new Map();
themeMap.set('sap_belize', sapBelize);
themeMap.set('sap_belize_plus', sapBelizePlus);
themeMap.set('sap_fiori_3', sapFiori3);
themeMap.set('sap_fiori_3_dark', sapFiori3Dark);
themeMap.set('sap_belize_hcw', sapBelizeHCW);
themeMap.set('sap_belize_hcb', sapBelizeHCB);
var getOrLoadTheme = function getOrLoadTheme(theme) {
  if (themeMap.has(theme)) {
    return themeMap.get(theme);
  } else {
    console.warn("".concat(theme, "-theme not found. Falling back to default theme 'sap_fiori_3'."));
    return themeMap.get('sap_fiori_3');
  }
};

var populateData = function populateData(labels, datasets, colors, theme) {
  var isPie = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : false;
  var colorPalette = resolveColors(colors, theme);
  return {
    labels: labels,
    datasets: datasets.map(function (item, index) {
      return _objectSpread({
        backgroundColor: isPie ? colorPalette : colorPalette[index % colorPalette.length],
        borderColor: isPie ? colorPalette : colorPalette[index % colorPalette.length]
      }, item);
    })
  };
};
var resolveColors = function resolveColors(colors, theme) {
  var _getOrLoadTheme = getOrLoadTheme(theme),
      semanticColors = _getOrLoadTheme.semanticColors,
      sequentialColors = _getOrLoadTheme.sequentialColors;

  if (colors && colors.length > 0) {
    return colors.map(function (key) {
      return semanticColors[key] || sequentialColors[key] || key;
    });
  }

  return Object.values(sequentialColors);
};
var useChartData = function useChartData(labels, datasets, colors, theme) {
  var isPie = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : false;
  return useMemo(function () {
    return populateData(labels, datasets, colors, theme, isPie);
  }, [labels, datasets, colors, theme, isPie]);
};

var useMergedConfig = function useMergedConfig(x, y) {
  return useMemo(function () {
    return merge(x, y);
  }, [x, y]);
}; // this needs to be a function as we need the `this` of the chart;

var formatAxisCallback = function formatAxisCallback(formatter) {
  return function (value) {
    // @ts-ignore
    var currentDataset = this.chart.data.datasets[0];
    return formatter(value, currentDataset);
  };
};
var formatTooltipLabel = function formatTooltipLabel(categoryFormatter, valueFormatter) {
  var valueAccessor = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'yLabel';
  return function (tooltipItem, data) {
    var currentDataset = data.datasets[tooltipItem.datasetIndex];
    var label = categoryFormatter(currentDataset.label, currentDataset, tooltipItem) || '';

    if (label) {
      label += ': ';
    }

    label += valueFormatter(tooltipItem[valueAccessor], currentDataset, tooltipItem);
    return label;
  };
};

var RadarChartComponent = forwardRef(function (props, ref) {
  var _ref = props,
      labels = _ref.labels,
      datasets = _ref.datasets,
      width = _ref.width,
      height = _ref.height,
      options = _ref.options,
      categoryAxisFormatter = _ref.categoryAxisFormatter,
      getDatasetAtEvent = _ref.getDatasetAtEvent,
      getElementAtEvent = _ref.getElementAtEvent,
      valueAxisFormatter = _ref.valueAxisFormatter,
      colors = _ref.colors,
      noLegend = _ref.noLegend,
      legendRef = _ref.legendRef;
  useEffect(function () {
    deprecationNotice('RadarChart', "This component is deprecated and will be removed with v0.10.0. Please use '@ui5/webcomponents-react-charts/lib/next/RadarChart' instead.");
  }, []);
  var theme = getTheme();
  var data = useChartData(labels, datasets, colors, theme);
  var radarChartDefaultConfig = useMemo(function () {
    return {
      scale: {
        ticks: {
          callback: formatAxisCallback(valueAxisFormatter)
        }
      },
      tooltips: {
        callbacks: {
          label: formatTooltipLabel(categoryAxisFormatter, valueAxisFormatter)
        }
      },
      plugins: {
        datalabels: false
      }
    };
  }, [categoryAxisFormatter, valueAxisFormatter]);
  var mergedOptions = useMergedConfig(radarChartDefaultConfig, options);
  var chartRef = useConsolidatedRef(ref);
  var handleLegendItemPress = useLegendItemClickHandler(chartRef, legendRef);
  useLegend(chartRef, legendRef, noLegend, handleLegendItemPress);
  return /*#__PURE__*/React.createElement(Radar, {
    ref: chartRef,
    data: data,
    height: height,
    width: width,
    options: mergedOptions,
    getDatasetAtEvent: getDatasetAtEvent,
    getElementAtEvent: getElementAtEvent
  });
});
/**
 * <code>import { RadarChart } from '@ui5/webcomponents-react-charts/lib/RadarChart';</code>
 * <br />
 * <b>This component is deprecated and will be removed with v0.10.0. Please use [this component](https://sap.github.io/ui5-webcomponents-react/?path=/docs/charts-radarchart) instead.</b>
 */

var RadarChart = withChartContainer(RadarChartComponent);
RadarChart.defaultProps = _objectSpread({}, ChartBaseDefaultProps);
RadarChart.displayName = 'RadarChart';

export { RadarChart };
