// @ui5/webcomponents-react-charts/lib/next/RadialChart.js
import { enrichEventWithDetails } from '@ui5/webcomponents-react-base/lib/Utils';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';
import { ChartContainer } from '@ui5/webcomponents-react-charts/lib/next/ChartContainer';
import { PieChartPlaceholder } from '@ui5/webcomponents-react-charts/lib/PieChartPlaceholder';
import React, { forwardRef, useMemo, useCallback } from 'react';
import { RadialBarChart, PolarAngleAxis, RadialBar } from 'recharts';

var radialChartMargin = {
  right: 30,
  left: 30,
  top: 30,
  bottom: 30
};
var radialBarBackground = {
  fill: ThemingParameters.sapContent_ImagePlaceholderBackground
};
var radialBarLabelStyle = {
  fontSize: ThemingParameters.sapFontHeader3Size,
  fill: ThemingParameters.sapTextColor
};
/**
 * <code>import { RadialChart } from '@ui5/webcomponents-react-charts/lib/next/RadialChart';</code>
 */

var RadialChart = forwardRef(function (props, ref) {
  var _props$maxValue = props.maxValue,
      maxValue = _props$maxValue === void 0 ? 100 : _props$maxValue,
      value = props.value,
      displayValue = props.displayValue,
      onDataPointClick = props.onDataPointClick,
      color = props.color,
      style = props.style,
      className = props.className,
      tooltip = props.tooltip,
      slot = props.slot;
  var range = useMemo(function () {
    return [0, maxValue];
  }, [maxValue]);
  var dataset = useMemo(function () {
    return [{
      value: value
    }];
  }, [value]);
  var onDataPointClickInternal = useCallback(function (payload, i, event) {
    if (payload && onDataPointClick) {
      onDataPointClick(enrichEventWithDetails(event, {
        value: payload.value,
        payload: payload.payload,
        dataIndex: i
      }));
    }
  }, [onDataPointClick]);
  return /*#__PURE__*/React.createElement(ChartContainer, {
    dataset: dataset,
    ref: ref,
    Placeholder: PieChartPlaceholder,
    style: style,
    className: className,
    tooltip: tooltip,
    slot: slot,
    resizeDebounce: 250
  }, /*#__PURE__*/React.createElement(RadialBarChart, {
    margin: radialChartMargin,
    innerRadius: "90%",
    outerRadius: "100%",
    barSize: 10,
    data: dataset,
    cx: "50%",
    cy: "50%",
    startAngle: 90,
    endAngle: -270
  }, /*#__PURE__*/React.createElement(PolarAngleAxis, {
    type: "number",
    domain: range,
    tick: false
  }), /*#__PURE__*/React.createElement(RadialBar, {
    background: radialBarBackground,
    dataKey: "value",
    cornerRadius: "50%",
    fill: color !== null && color !== void 0 ? color : ThemingParameters.sapChart_OrderedColor_1,
    onClick: onDataPointClickInternal
  }), /*#__PURE__*/React.createElement("text", {
    x: "50%",
    y: "50%",
    textAnchor: "middle",
    dominantBaseline: "middle",
    className: "progress-label",
    style: radialBarLabelStyle
  }, displayValue)));
});
RadialChart.displayName = 'RadialChart';

export { RadialChart };
