// @ui5/webcomponents-react-base/lib/sap_fiori_3.js
import { deprecationNotice } from '@ui5/webcomponents-react-base/lib/Utils';

/*
 * ### WARNING ###
 * This is an autogenerated file, do not change manually.
 * In order to recreate this file, please run 'node packages/base/scripts/cssVariables/parse.js'
 */

/* eslint-disable camelcase */

deprecationNotice('sap_fiori_3', "'@ui5/webcomponents-react-base/lib/sap_fiori_3' is deprecated and will be removed in the next major release.\n. Please use the '@ui5/webcomponents-react-base/lib/ThemingParameters' instead.");
var sapFontFamily = 'var(--sapFontFamily)';
var sapFontSize = 'var(--sapFontSize)';
var sapCompanyLogo = 'var(--sapCompanyLogo)';
var sapBackgroundImage = 'var(--sapBackgroundImage)';
var sapBackgroundImageOpacity = 'var(--sapBackgroundImageOpacity)';
var sapBackgroundImageRepeat = 'var(--sapBackgroundImageRepeat)';
var sapContent_GridSize = 'var(--sapContent_GridSize)';
var sapPrimary1 = 'var(--sapPrimary1)';
var sapPrimary2 = 'var(--sapPrimary2)';
var sapPrimary3 = 'var(--sapPrimary3)';
var sapPrimary4 = 'var(--sapPrimary4)';
var sapPrimary5 = 'var(--sapPrimary5)';
var sapPrimary6 = 'var(--sapPrimary6)';
var sapPrimary7 = 'var(--sapPrimary7)';
var sapAccentColor1 = 'var(--sapAccentColor1)';
var sapAccentColor2 = 'var(--sapAccentColor2)';
var sapAccentColor3 = 'var(--sapAccentColor3)';
var sapAccentColor4 = 'var(--sapAccentColor4)';
var sapAccentColor5 = 'var(--sapAccentColor5)';
var sapAccentColor6 = 'var(--sapAccentColor6)';
var sapAccentColor7 = 'var(--sapAccentColor7)';
var sapAccentColor8 = 'var(--sapAccentColor8)';
var sapAccentColor9 = 'var(--sapAccentColor9)';
var sapAccentColor10 = 'var(--sapAccentColor10)';
var sapNegativeColor = 'var(--sapNegativeColor)';
var sapCriticalColor = 'var(--sapCriticalColor)';
var sapPositiveColor = 'var(--sapPositiveColor)';
var sapInformativeColor = 'var(--sapInformativeColor)';
var sapNeutralColor = 'var(--sapNeutralColor)';
var sapIndicationColor_1 = 'var(--sapIndicationColor_1)';
var sapIndicationColor_2 = 'var(--sapIndicationColor_2)';
var sapIndicationColor_3 = 'var(--sapIndicationColor_3)';
var sapIndicationColor_4 = 'var(--sapIndicationColor_4)';
var sapIndicationColor_5 = 'var(--sapIndicationColor_5)';
var sapElement_LineHeight = 'var(--sapElement_LineHeight)';
var sapElement_Height = 'var(--sapElement_Height)';
var sapElement_BorderWidth = 'var(--sapElement_BorderWidth)';
var sapContent_LineHeight = 'var(--sapContent_LineHeight)';
var sapContent_ElementHeight = 'var(--sapContent_ElementHeight)';
var sapContent_ElementHeight_PX = 'var(--sapContent_ElementHeight_PX)';
var sapContent_FocusColor = 'var(--sapContent_FocusColor)';
var sapContent_ContrastFocusColor = 'var(--sapContent_ContrastFocusColor)';
var sapContent_ShadowColor = 'var(--sapContent_ShadowColor)';
var sapContent_ContrastShadowColor = 'var(--sapContent_ContrastShadowColor)';
var sapContent_SearchHighlightColor = 'var(--sapContent_SearchHighlightColor)';
var sapContent_HelpColor = 'var(--sapContent_HelpColor)';
var sapContent_MonospaceFontFamily = 'var(--sapContent_MonospaceFontFamily)';
var sapContent_DisabledOpacity = 'var(--sapContent_DisabledOpacity)';
var sapContent_ContrastTextThreshold = 'var(--sapContent_ContrastTextThreshold)';
var sapContent_ContrastTextColor = 'var(--sapContent_ContrastTextColor)';
var sapShell_BackgroundImage = 'var(--sapShell_BackgroundImage)';
var sapShell_BackgroundImageOpacity = 'var(--sapShell_BackgroundImageOpacity)';
var sapShell_BackgroundImageRepeat = 'var(--sapShell_BackgroundImageRepeat)';
var sapShell_BackgroundPatternColor = 'var(--sapShell_BackgroundPatternColor)';
var sapShell_BackgroundGradient = 'var(--sapShell_BackgroundGradient)';
var sapShell_InteractiveTextColor = 'var(--sapShell_InteractiveTextColor)';
var sapShell_Favicon = 'var(--sapShell_Favicon)';
var sapButton_BorderCornerRadius = 'var(--sapButton_BorderCornerRadius)';
var sapField_BorderCornerRadius = 'var(--sapField_BorderCornerRadius)';
var sapGroup_TitleBackground = 'var(--sapGroup_TitleBackground)';
var sapGroup_BorderCornerRadius = 'var(--sapGroup_BorderCornerRadius)';
var sapGroup_FooterBackground = 'var(--sapGroup_FooterBackground)';
var sapToolbar_Background = 'var(--sapToolbar_Background)';
var sapScrollBar_Dimension = 'var(--sapScrollBar_Dimension)';
var sapBlockLayer_Background = 'var(--sapBlockLayer_Background)';
var sapTile_BorderColor = 'var(--sapTile_BorderColor)';
var sapUiFontSize = 'var(--sapUiFontSize)';
var sapUiFontCondensedFamily = 'var(--sapUiFontCondensedFamily)';
var sapUiFontHeaderWeight = 'var(--sapUiFontHeaderWeight)';
var sapMFontHeader1Size = 'var(--sapMFontHeader1Size)';
var sapMFontHeader2Size = 'var(--sapMFontHeader2Size)';
var sapMFontHeader3Size = 'var(--sapMFontHeader3Size)';
var sapMFontHeader4Size = 'var(--sapMFontHeader4Size)';
var sapMFontHeader5Size = 'var(--sapMFontHeader5Size)';
var sapMFontHeader6Size = 'var(--sapMFontHeader6Size)';
var sapMFontSmallSize = 'var(--sapMFontSmallSize)';
var sapMFontMediumSize = 'var(--sapMFontMediumSize)';
var sapMFontLargeSize = 'var(--sapMFontLargeSize)';
var sapUiTranslucentBGOpacity = 'var(--sapUiTranslucentBGOpacity)';
var sapMPlatformDependent = 'var(--sapMPlatformDependent)';
var sapUiDesktopFontSize = 'var(--sapUiDesktopFontSize)';
var sapUiFontLargeSize = 'var(--sapUiFontLargeSize)';
var sapUiFontSmallSize = 'var(--sapUiFontSmallSize)';
var sapUiFontHeader1Size = 'var(--sapUiFontHeader1Size)';
var sapUiFontHeader2Size = 'var(--sapUiFontHeader2Size)';
var sapUiFontHeader3Size = 'var(--sapUiFontHeader3Size)';
var sapUiFontHeader4Size = 'var(--sapUiFontHeader4Size)';
var sapUiFontHeader5Size = 'var(--sapUiFontHeader5Size)';
var sapUiLineHeight = 'var(--sapUiLineHeight)';
var sapUiNotifierSeparatorWidth = 'var(--sapUiNotifierSeparatorWidth)';
var sapUiButtonEmphasizedTextShadow = 'var(--sapUiButtonEmphasizedTextShadow)';
var sapUiButtonLiteBackground = 'var(--sapUiButtonLiteBackground)';
var sapUiButtonLiteBorderColor = 'var(--sapUiButtonLiteBorderColor)';
var sapUiListTableTextSize = 'var(--sapUiListTableTextSize)';
var sapUiListTableIconSize = 'var(--sapUiListTableIconSize)';
var sap_wc_input_icon_min_width = 'var(--sap_wc_input_icon_min_width)';
var sap_wc_input_compact_min_width = 'var(--sap_wc_input_compact_min_width)';
var sap_wc_input_disabled_opacity = 'var(--sap_wc_input_disabled_opacity)';
var sapShellColor = 'var(--sapShellColor)';
var sapBrandColor = 'var(--sapBrandColor)';
var sapLinkColor = 'var(--sapLinkColor)';
var sapBaseColor = 'var(--sapBaseColor)';
var sapTextColor = 'var(--sapTextColor)';
var sapErrorBorderColor = 'var(--sapErrorBorderColor)';
var sapNegativeElementColor = 'var(--sapNegativeElementColor)';
var sapNegativeTextColor = 'var(--sapNegativeTextColor)';
var sapWarningBorderColor = 'var(--sapWarningBorderColor)';
var sapCriticalElementColor = 'var(--sapCriticalElementColor)';
var sapCriticalTextColor = 'var(--sapCriticalTextColor)';
var sapSuccessBorderColor = 'var(--sapSuccessBorderColor)';
var sapPositiveElementColor = 'var(--sapPositiveElementColor)';
var sapPositiveTextColor = 'var(--sapPositiveTextColor)';
var sapInformationBorderColor = 'var(--sapInformationBorderColor)';
var sapInformativeElementColor = 'var(--sapInformativeElementColor)';
var sapNeutralBorderColor = 'var(--sapNeutralBorderColor)';
var sapNeutralElementColor = 'var(--sapNeutralElementColor)';
var sapNeutralTextColor = 'var(--sapNeutralTextColor)';
var sapErrorColor = 'var(--sapErrorColor)';
var sapWarningColor = 'var(--sapWarningColor)';
var sapSuccessColor = 'var(--sapSuccessColor)';
var sapInformationColor = 'var(--sapInformationColor)';
var sapContent_IconHeight = 'var(--sapContent_IconHeight)';
var sapContent_NonInteractiveIconColor = 'var(--sapContent_NonInteractiveIconColor)';
var sapContent_LabelColor = 'var(--sapContent_LabelColor)';
var sapContent_ContrastIconColor = 'var(--sapContent_ContrastIconColor)';
var sapContent_ForegroundBorderColor = 'var(--sapContent_ForegroundBorderColor)';
var sapShell_Background = 'var(--sapShell_Background)';
var sapButton_BorderWidth = 'var(--sapButton_BorderWidth)';
var sapField_BorderColor = 'var(--sapField_BorderColor)';
var sapField_BorderWidth = 'var(--sapField_BorderWidth)';
var sapGroup_Title_FontSize = 'var(--sapGroup_Title_FontSize)';
var sapGroup_BorderWidth = 'var(--sapGroup_BorderWidth)';
var sapList_BorderWidth = 'var(--sapList_BorderWidth)';
var sapUiFontFamily = 'var(--sapUiFontFamily)';
var sapUiGlobalLogo = 'var(--sapUiGlobalLogo)';
var sapUiGlobalBackgroundImage = 'var(--sapUiGlobalBackgroundImage)';
var sapUiBackgroundImage = 'var(--sapUiBackgroundImage)';
var sapUiUx3ShellBackgroundImageURL = 'var(--sapUiUx3ShellBackgroundImageURL)';
var sapUiGlobalBackgroundImageOpacity = 'var(--sapUiGlobalBackgroundImageOpacity)';
var sapUiGlobalBackgroundRepeat = 'var(--sapUiGlobalBackgroundRepeat)';
var sapUiElementLineHeight = 'var(--sapUiElementLineHeight)';
var sapUiElementHeight = 'var(--sapUiElementHeight)';
var sapUiElementBorderWidth = 'var(--sapUiElementBorderWidth)';
var sapUiContentLineHeight = 'var(--sapUiContentLineHeight)';
var sapUiContentElementHeight = 'var(--sapUiContentElementHeight)';
var sapUiContentElementHeightPX = 'var(--sapUiContentElementHeightPX)';
var sapUiContentFocusColor = 'var(--sapUiContentFocusColor)';
var sapUiContentContrastFocusColor = 'var(--sapUiContentContrastFocusColor)';
var sapUiContentShadowColor = 'var(--sapUiContentShadowColor)';
var sapUiContentContrastShadowColor = 'var(--sapUiContentContrastShadowColor)';
var sapUiContentSearchHighlightColor = 'var(--sapUiContentSearchHighlightColor)';
var sapUiContentHelpColor = 'var(--sapUiContentHelpColor)';
var sapUiContentDisabledOpacity = 'var(--sapUiContentDisabledOpacity)';
var sapUiContentContrastTextThreshold = 'var(--sapUiContentContrastTextThreshold)';
var sapUiContentContrastTextColor = 'var(--sapUiContentContrastTextColor)';
var sapUiShellBackgroundImage = 'var(--sapUiShellBackgroundImage)';
var sapUiShellBackgroundImageOpacity = 'var(--sapUiShellBackgroundImageOpacity)';
var sapUiShellBackgroundImageRepeat = 'var(--sapUiShellBackgroundImageRepeat)';
var sapUiShellBackgroundPatternColor = 'var(--sapUiShellBackgroundPatternColor)';
var sapUiShellBackgroundGradient = 'var(--sapUiShellBackgroundGradient)';
var sapUiShellInteractiveTextColor = 'var(--sapUiShellInteractiveTextColor)';
var sapUiShellFavicon = 'var(--sapUiShellFavicon)';
var sapUiButtonBorderCornerRadius = 'var(--sapUiButtonBorderCornerRadius)';
var sapUiFieldBorderCornerRadius = 'var(--sapUiFieldBorderCornerRadius)';
var sapUiGroupTitleBackground = 'var(--sapUiGroupTitleBackground)';
var sapUiGroupFooterBackground = 'var(--sapUiGroupFooterBackground)';
var sapUiToolbarBackground = 'var(--sapUiToolbarBackground)';
var sapUiScrollBarDimension = 'var(--sapUiScrollBarDimension)';
var sapUiBlockLayerBackground = 'var(--sapUiBlockLayerBackground)';
var sapUiTileBorderColor = 'var(--sapUiTileBorderColor)';
var sapUiContentGridSize = 'var(--sapUiContentGridSize)';
var sapUiPrimary1 = 'var(--sapUiPrimary1)';
var sapUiPrimary2 = 'var(--sapUiPrimary2)';
var sapUiPrimary3 = 'var(--sapUiPrimary3)';
var sapUiPrimary4 = 'var(--sapUiPrimary4)';
var sapUiPrimary5 = 'var(--sapUiPrimary5)';
var sapUiPrimary6 = 'var(--sapUiPrimary6)';
var sapUiPrimary7 = 'var(--sapUiPrimary7)';
var sapUiAccent1 = 'var(--sapUiAccent1)';
var sapUiAccent2 = 'var(--sapUiAccent2)';
var sapUiAccent3 = 'var(--sapUiAccent3)';
var sapUiAccent4 = 'var(--sapUiAccent4)';
var sapUiAccent5 = 'var(--sapUiAccent5)';
var sapUiAccent6 = 'var(--sapUiAccent6)';
var sapUiAccent7 = 'var(--sapUiAccent7)';
var sapUiAccent8 = 'var(--sapUiAccent8)';
var sapUiAccent9 = 'var(--sapUiAccent9)';
var sapUiAccent10 = 'var(--sapUiAccent10)';
var sapUiNeutralBorder = 'var(--sapUiNeutralBorder)';
var sapUiNegative = 'var(--sapUiNegative)';
var sapUiCritical = 'var(--sapUiCritical)';
var sapUiPositive = 'var(--sapUiPositive)';
var sapUiInformative = 'var(--sapUiInformative)';
var sapUiNeutral = 'var(--sapUiNeutral)';
var sapUiIndication1 = 'var(--sapUiIndication1)';
var sapUiIndication2 = 'var(--sapUiIndication2)';
var sapUiIndication3 = 'var(--sapUiIndication3)';
var sapUiIndication4 = 'var(--sapUiIndication4)';
var sapUiIndication5 = 'var(--sapUiIndication5)';
var sapUiFontHeader6Size = 'var(--sapUiFontHeader6Size)';
var sapShell_BorderColor = 'var(--sapShell_BorderColor)';
var sapUiShellColor = 'var(--sapUiShellColor)';
var sapButton_Emphasized_Background = 'var(--sapButton_Emphasized_Background)';
var sapUiBrand = 'var(--sapUiBrand)';
var sapUiLink = 'var(--sapUiLink)';
var sapContent_ImagePlaceholderForegroundColor = 'var(--sapContent_ImagePlaceholderForegroundColor)';
var sapButton_Background = 'var(--sapButton_Background)';
var sapField_Background = 'var(--sapField_Background)';
var sapGroup_ContentBackground = 'var(--sapGroup_ContentBackground)';
var sapList_Background = 'var(--sapList_Background)';
var sapPageHeader_Background = 'var(--sapPageHeader_Background)';
var sapPageFooter_Background = 'var(--sapPageFooter_Background)';
var sapObjectHeader_Background = 'var(--sapObjectHeader_Background)';
var sapTile_Background = 'var(--sapTile_Background)';
var sapUiBaseColor = 'var(--sapUiBaseColor)';
var sapTitleColor = 'var(--sapTitleColor)';
var sapField_TextColor = 'var(--sapField_TextColor)';
var sapUiBaseText = 'var(--sapUiBaseText)';
var sapField_InvalidColor = 'var(--sapField_InvalidColor)';
var sapUiErrorBorder = 'var(--sapUiErrorBorder)';
var sapUiNegativeElement = 'var(--sapUiNegativeElement)';
var sapUiNegativeText = 'var(--sapUiNegativeText)';
var sapField_WarningColor = 'var(--sapField_WarningColor)';
var sapUiWarningBorder = 'var(--sapUiWarningBorder)';
var sapUiCriticalElement = 'var(--sapUiCriticalElement)';
var sapUiCriticalText = 'var(--sapUiCriticalText)';
var sapField_SuccessColor = 'var(--sapField_SuccessColor)';
var sapUiSuccessBorder = 'var(--sapUiSuccessBorder)';
var sapUiPositiveElement = 'var(--sapUiPositiveElement)';
var sapUiPositiveText = 'var(--sapUiPositiveText)';
var sapUiInformationBorder = 'var(--sapUiInformationBorder)';
var sapUiInformativeElement = 'var(--sapUiInformativeElement)';
var sapUiNeutralElement = 'var(--sapUiNeutralElement)';
var sapUiNeutralText = 'var(--sapUiNeutralText)';
var sapUiContentIconHeight = 'var(--sapUiContentIconHeight)';
var sapUiContentNonInteractiveIconColor = 'var(--sapUiContentNonInteractiveIconColor)';
var sapUiContentLabelColor = 'var(--sapUiContentLabelColor)';
var sapUiContentContrastIconColor = 'var(--sapUiContentContrastIconColor)';
var sapUiContentForegroundBorderColor = 'var(--sapUiContentForegroundBorderColor)';
var sapUiShellBackground = 'var(--sapUiShellBackground)';
var sapUiButtonBorderWidth = 'var(--sapUiButtonBorderWidth)';
var sapField_Focus_BorderColor = 'var(--sapField_Focus_BorderColor)';
var sapField_ReadOnly_BorderColor = 'var(--sapField_ReadOnly_BorderColor)';
var sapUiFieldBorderColor = 'var(--sapUiFieldBorderColor)';
var sapUiFieldBorderWidth = 'var(--sapUiFieldBorderWidth)';
var sapUiGroupBorderWidth = 'var(--sapUiGroupBorderWidth)';
var sapUiListBorderWidth = 'var(--sapUiListBorderWidth)';
var sapUiFontHeaderFamily = 'var(--sapUiFontHeaderFamily)';
var sapUiDesktopFontFamily = 'var(--sapUiDesktopFontFamily)';
var sapUiShadowText = 'var(--sapUiShadowText)';
var sapUiFieldActiveTextColor = 'var(--sapUiFieldActiveTextColor)';
var sapUiCalendarColorToday = 'var(--sapUiCalendarColorToday)';
var sapUiShellBorderColor = 'var(--sapUiShellBorderColor)';
var sapButton_Emphasized_BorderColor = 'var(--sapButton_Emphasized_BorderColor)';
var sapUiButtonEmphasizedBackground = 'var(--sapUiButtonEmphasizedBackground)';
var sapUiLinkActive = 'var(--sapUiLinkActive)';
var sapUiLinkVisited = 'var(--sapUiLinkVisited)';
var sapUiDragAndDropActiveColor = 'var(--sapUiDragAndDropActiveColor)';
var sapUiDragAndDropActiveBorderColor = 'var(--sapUiDragAndDropActiveBorderColor)';
var sapUiContentImagePlaceholderForegroundColor = 'var(--sapUiContentImagePlaceholderForegroundColor)';
var sapButton_Reject_Background = 'var(--sapButton_Reject_Background)';
var sapButton_Accept_Background = 'var(--sapButton_Accept_Background)';
var sapUiButtonBackground = 'var(--sapUiButtonBackground)';
var sapField_HelpBackground = 'var(--sapField_HelpBackground)';
var sapField_Hover_Background = 'var(--sapField_Hover_Background)';
var sapField_Focus_Background = 'var(--sapField_Focus_Background)';
var sapField_InvalidBackground = 'var(--sapField_InvalidBackground)';
var sapField_WarningBackground = 'var(--sapField_WarningBackground)';
var sapField_SuccessBackground = 'var(--sapField_SuccessBackground)';
var sapUiFieldBackground = 'var(--sapUiFieldBackground)';
var sapUiGroupContentBackground = 'var(--sapUiGroupContentBackground)';
var sapUiListBackground = 'var(--sapUiListBackground)';
var sapUiPageHeaderBackground = 'var(--sapUiPageHeaderBackground)';
var sapUiPageFooterBackground = 'var(--sapUiPageFooterBackground)';
var sapUiObjectHeaderBackground = 'var(--sapUiObjectHeaderBackground)';
var sapUiTileBackground = 'var(--sapUiTileBackground)';
var sapUiTextTitle = 'var(--sapUiTextTitle)';
var sapUiFieldTextColor = 'var(--sapUiFieldTextColor)';
var sapUiListTextColor = 'var(--sapUiListTextColor)';
var sapUiFieldInvalidColor = 'var(--sapUiFieldInvalidColor)';
var sapUiButtonRejectBorderColor = 'var(--sapUiButtonRejectBorderColor)';
var sapUiFieldWarningColor = 'var(--sapUiFieldWarningColor)';
var sapUiFieldSuccessColor = 'var(--sapUiFieldSuccessColor)';
var sapUiButtonAcceptBorderColor = 'var(--sapUiButtonAcceptBorderColor)';
var sapUiDragAndDropColor = 'var(--sapUiDragAndDropColor)';
var sapUiDragAndDropBorderColor = 'var(--sapUiDragAndDropBorderColor)';
var sapUiFieldFocusBorderColor = 'var(--sapUiFieldFocusBorderColor)';
var sapUiFieldReadOnlyBorderColor = 'var(--sapUiFieldReadOnlyBorderColor)';
var sapUiButtonEmphasizedBorderColor = 'var(--sapUiButtonEmphasizedBorderColor)';
var sapUiButtonRejectBackground = 'var(--sapUiButtonRejectBackground)';
var sapUiButtonAcceptBackground = 'var(--sapUiButtonAcceptBackground)';
var sapUiSegmentedButtonBackground = 'var(--sapUiSegmentedButtonBackground)';
var sapUiFieldHelpBackground = 'var(--sapUiFieldHelpBackground)';
var sapUiFieldHoverBackground = 'var(--sapUiFieldHoverBackground)';
var sapField_Focus_HelpBackground = 'var(--sapField_Focus_HelpBackground)';
var sapUiFieldFocusBackground = 'var(--sapUiFieldFocusBackground)';
var sapUiFieldInvalidBackground = 'var(--sapUiFieldInvalidBackground)';
var sapUiFieldWarningBackground = 'var(--sapUiFieldWarningBackground)';
var sapUiFieldSuccessBackground = 'var(--sapUiFieldSuccessBackground)';
var sapUiShellContainerBackground = 'var(--sapUiShellContainerBackground)';
var sapUiDragAndDropBackground = 'var(--sapUiDragAndDropBackground)';
var sapUiButtonActionSelectBackground = 'var(--sapUiButtonActionSelectBackground)';
var sapUiListGroupHeaderBackground = 'var(--sapUiListGroupHeaderBackground)';
var sapUiShellGroupTextColor = 'var(--sapUiShellGroupTextColor)';
var sapUiListFooterTextColor = 'var(--sapUiListFooterTextColor)';
var sapUiButtonRejectHoverBorderColor = 'var(--sapUiButtonRejectHoverBorderColor)';
var sapUiButtonAcceptHoverBorderColor = 'var(--sapUiButtonAcceptHoverBorderColor)';
var sapUiFieldFocusHelpBackground = 'var(--sapUiFieldFocusHelpBackground)';
var sapUiShellAltContainerBackground = 'var(--sapUiShellAltContainerBackground)';
var sapHighlightColor = 'var(--sapHighlightColor)';
var sapBackgroundColorDefault = 'var(--sapBackgroundColorDefault)';
var sapBackgroundColor = 'var(--sapBackgroundColor)';
var sapErrorBackground = 'var(--sapErrorBackground)';
var sapWarningBackground = 'var(--sapWarningBackground)';
var sapSuccessBackground = 'var(--sapSuccessBackground)';
var sapInformationBackground = 'var(--sapInformationBackground)';
var sapNeutralBackground = 'var(--sapNeutralBackground)';
var sapInformativeTextColor = 'var(--sapInformativeTextColor)';
var sapContent_MarkerIconColor = 'var(--sapContent_MarkerIconColor)';
var sapContent_MarkerTextColor = 'var(--sapContent_MarkerTextColor)';
var sapContent_ImagePlaceholderBackground = 'var(--sapContent_ImagePlaceholderBackground)';
var sapContent_DisabledTextColor = 'var(--sapContent_DisabledTextColor)';
var sapContent_ForegroundColor = 'var(--sapContent_ForegroundColor)';
var sapContent_BadgeBackground = 'var(--sapContent_BadgeBackground)';
var sapButton_BorderColor = 'var(--sapButton_BorderColor)';
var sapField_RequiredColor = 'var(--sapField_RequiredColor)';
var sapGroup_TitleBorderColor = 'var(--sapGroup_TitleBorderColor)';
var sapScrollBar_FaceColor = 'var(--sapScrollBar_FaceColor)';
var sapInfobar_Background = 'var(--sapInfobar_Background)';
var sapToolbar_SeparatorColor = 'var(--sapToolbar_SeparatorColor)';
var sapUiNotificationBarBG = 'var(--sapUiNotificationBarBG)';
var sapUiNotifierSeparator = 'var(--sapUiNotifierSeparator)';
var sapUiNotificationBarBorder = 'var(--sapUiNotificationBarBorder)';
var sapUiContentShadowColorFade30 = 'var(--sapUiContentShadowColorFade30)';
var sapUiContentShadowColorFade5 = 'var(--sapUiContentShadowColorFade5)';
var sapUiContentShadowColorFade15 = 'var(--sapUiContentShadowColorFade15)';
var sapUiAccent1Lighten50 = 'var(--sapUiAccent1Lighten50)';
var sapUiAccent2Lighten40 = 'var(--sapUiAccent2Lighten40)';
var sapUiAccent3Lighten46 = 'var(--sapUiAccent3Lighten46)';
var sapUiAccent4Lighten46 = 'var(--sapUiAccent4Lighten46)';
var sapUiAccent5Lighten32 = 'var(--sapUiAccent5Lighten32)';
var sapUiAccent6Lighten52 = 'var(--sapUiAccent6Lighten52)';
var sapUiAccent7Lighten64 = 'var(--sapUiAccent7Lighten64)';
var sapUiAccent8Lighten61 = 'var(--sapUiAccent8Lighten61)';
var sapUiAccent9Lighten37 = 'var(--sapUiAccent9Lighten37)';
var sapUiAccent10Lighten49 = 'var(--sapUiAccent10Lighten49)';
var sapUiShellHoverBackground = 'var(--sapUiShellHoverBackground)';
var sapUiShellActiveBackground = 'var(--sapUiShellActiveBackground)';
var sapUiShellHoverToggleBackground = 'var(--sapUiShellHoverToggleBackground)';
var sapUiLinkHover = 'var(--sapUiLinkHover)';
var sapUiDragAndDropActiveBackground = 'var(--sapUiDragAndDropActiveBackground)';
var sapUiLinkDarken15 = 'var(--sapUiLinkDarken15)';
var sapGroup_ContentBorderColor = 'var(--sapGroup_ContentBorderColor)';
var sapList_BorderColor = 'var(--sapList_BorderColor)';
var sapList_HeaderBackground = 'var(--sapList_HeaderBackground)';
var sapPageHeader_BorderColor = 'var(--sapPageHeader_BorderColor)';
var sapUiButtonRejectActiveBackground = 'var(--sapUiButtonRejectActiveBackground)';
var sapUiButtonAcceptActiveBackground = 'var(--sapUiButtonAcceptActiveBackground)';
var sapUiFieldPlaceholderTextColor = 'var(--sapUiFieldPlaceholderTextColor)';
var sapSelectedColor = 'var(--sapSelectedColor)';
var sapActiveColor = 'var(--sapActiveColor)';
var sapContent_IconColor = 'var(--sapContent_IconColor)';
var sapField_Hover_BorderColor = 'var(--sapField_Hover_BorderColor)';
var sapList_HighlightColor = 'var(--sapList_HighlightColor)';
var sapUiHighlight = 'var(--sapUiHighlight)';
var sapUiBaseBG = 'var(--sapUiBaseBG)';
var sapUiGlobalBackgroundColor = 'var(--sapUiGlobalBackgroundColor)';
var sapUiErrorBG = 'var(--sapUiErrorBG)';
var sapUiWarningBG = 'var(--sapUiWarningBG)';
var sapUiSuccessBG = 'var(--sapUiSuccessBG)';
var sapUiInformationBG = 'var(--sapUiInformationBG)';
var sapUiNeutralBG = 'var(--sapUiNeutralBG)';
var sapUiInformativeText = 'var(--sapUiInformativeText)';
var sapUiContentMarkerIconColor = 'var(--sapUiContentMarkerIconColor)';
var sapUiContentMarkerTextColor = 'var(--sapUiContentMarkerTextColor)';
var sapUiContentImagePlaceholderBackground = 'var(--sapUiContentImagePlaceholderBackground)';
var sapUiContentDisabledTextColor = 'var(--sapUiContentDisabledTextColor)';
var sapUiContentForegroundColor = 'var(--sapUiContentForegroundColor)';
var sapUiContentBadgeBackground = 'var(--sapUiContentBadgeBackground)';
var sapButton_Hover_BorderColor = 'var(--sapButton_Hover_BorderColor)';
var sapUiButtonBorderColor = 'var(--sapUiButtonBorderColor)';
var sapUiFieldRequiredColor = 'var(--sapUiFieldRequiredColor)';
var sapUiGroupTitleBorderColor = 'var(--sapUiGroupTitleBorderColor)';
var sapScrollBar_BorderColor = 'var(--sapScrollBar_BorderColor)';
var sapUiScrollBarFaceColor = 'var(--sapUiScrollBarFaceColor)';
var sapUiInfobarBackground = 'var(--sapUiInfobarBackground)';
var sapUiToolbarSeparatorColor = 'var(--sapUiToolbarSeparatorColor)';
var sapUiShellBorderColorLighten30 = 'var(--sapUiShellBorderColorLighten30)';
var sapUiButtonEmphasizedHoverBackground = 'var(--sapUiButtonEmphasizedHoverBackground)';
var sapUiButtonEmphasizedActiveBackground = 'var(--sapUiButtonEmphasizedActiveBackground)';
var sapUiButtonBackgroundDarken24 = 'var(--sapUiButtonBackgroundDarken24)';
var sapUiButtonBackgroundDarken7 = 'var(--sapUiButtonBackgroundDarken7)';
var sapUiButtonBackgroundDarken2 = 'var(--sapUiButtonBackgroundDarken2)';
var sapUiButtonBackgroundDarken10 = 'var(--sapUiButtonBackgroundDarken10)';
var sapUiGroupContentBorderColor = 'var(--sapUiGroupContentBorderColor)';
var sapUiListFooterBackground = 'var(--sapUiListFooterBackground)';
var sapUiListTableGroupHeaderBackground = 'var(--sapUiListTableGroupHeaderBackground)';
var sapUiListBackgroundDarken3 = 'var(--sapUiListBackgroundDarken3)';
var sapUiListBackgroundDarken10 = 'var(--sapUiListBackgroundDarken10)';
var sapUiListBackgroundDarken13 = 'var(--sapUiListBackgroundDarken13)';
var sapUiListBackgroundDarken15 = 'var(--sapUiListBackgroundDarken15)';
var sapUiListBackgroundDarken20 = 'var(--sapUiListBackgroundDarken20)';
var sapList_HeaderBorderColor = 'var(--sapList_HeaderBorderColor)';
var sapUiListBorderColor = 'var(--sapUiListBorderColor)';
var sapUiListHeaderBackground = 'var(--sapUiListHeaderBackground)';
var sapUiPageHeaderBorderColor = 'var(--sapUiPageHeaderBorderColor)';
var sapUiObjectHeaderBorderColor = 'var(--sapUiObjectHeaderBorderColor)';
var sapUiTileBackgroundDarken20 = 'var(--sapUiTileBackgroundDarken20)';
var sapUiButtonRejectActiveBorderColor = 'var(--sapUiButtonRejectActiveBorderColor)';
var sapUiFieldWarningColorDarken100 = 'var(--sapUiFieldWarningColorDarken100)';
var sapUiButtonAcceptActiveBorderColor = 'var(--sapUiButtonAcceptActiveBorderColor)';
var sapUiSelected = 'var(--sapUiSelected)';
var sapUiActive = 'var(--sapUiActive)';
var sapScrollBar_SymbolColor = 'var(--sapScrollBar_SymbolColor)';
var sapUiContentIconColor = 'var(--sapUiContentIconColor)';
var sapUiFieldHoverBorderColor = 'var(--sapUiFieldHoverBorderColor)';
var sapUiListHighlightColor = 'var(--sapUiListHighlightColor)';
var sapUiButtonRejectHoverBackground = 'var(--sapUiButtonRejectHoverBackground)';
var sapUiButtonAcceptHoverBackground = 'var(--sapUiButtonAcceptHoverBackground)';
var sapUiButtonHoverBorderColor = 'var(--sapUiButtonHoverBorderColor)';
var sapUiSegmentedButtonBorderColor = 'var(--sapUiSegmentedButtonBorderColor)';
var sapUiScrollBarBorderColor = 'var(--sapUiScrollBarBorderColor)';
var sapUiButtonEmphasizedHoverBorderColor = 'var(--sapUiButtonEmphasizedHoverBorderColor)';
var sapUiButtonEmphasizedActiveBorderColor = 'var(--sapUiButtonEmphasizedActiveBorderColor)';
var sapUiListHeaderBorderColor = 'var(--sapUiListHeaderBorderColor)';
var sapUiButtonActionSelectBorderColor = 'var(--sapUiButtonActionSelectBorderColor)';
var sapUiListVerticalBorderColor = 'var(--sapUiListVerticalBorderColor)';
var sapUiListTableFooterBorder = 'var(--sapUiListTableFooterBorder)';
var sapUiToggleButtonPressedBackground = 'var(--sapUiToggleButtonPressedBackground)';
var sapUiButtonActiveBackground = 'var(--sapUiButtonActiveBackground)';
var sapUiFieldActiveBackground = 'var(--sapUiFieldActiveBackground)';
var sapUiFieldActiveBorderColor = 'var(--sapUiFieldActiveBorderColor)';
var sapUiScrollBarSymbolColor = 'var(--sapUiScrollBarSymbolColor)';
var sapUiListActiveBackground = 'var(--sapUiListActiveBackground)';
var sapUiSegmentedButtonFooterBorderColor = 'var(--sapUiSegmentedButtonFooterBorderColor)';
var sapUiToggleButtonPressedBorderColor = 'var(--sapUiToggleButtonPressedBorderColor)';
var sapUiSegmentedButtonSelectedBackground = 'var(--sapUiSegmentedButtonSelectedBackground)';
var sapUiButtonActiveBorderColor = 'var(--sapUiButtonActiveBorderColor)';
var sapUiButtonLiteActiveBackground = 'var(--sapUiButtonLiteActiveBackground)';
var sapUiSegmentedButtonActiveBackground = 'var(--sapUiSegmentedButtonActiveBackground)';
var sapUiInfobarActiveBackground = 'var(--sapUiInfobarActiveBackground)';
var sapUiLinkInverted = 'var(--sapUiLinkInverted)';
var sapField_ReadOnly_Background = 'var(--sapField_ReadOnly_Background)';
var sapUiButtonLiteActiveBorderColor = 'var(--sapUiButtonLiteActiveBorderColor)';
var sapField_ReadOnly_HelpBackground = 'var(--sapField_ReadOnly_HelpBackground)';
var sapUiFieldReadOnlyBackground = 'var(--sapUiFieldReadOnlyBackground)';
var sapButton_Hover_Background = 'var(--sapButton_Hover_Background)';
var sapBackgroundColorFade72 = 'var(--sapBackgroundColorFade72)';
var sapScrollBar_TrackColor = 'var(--sapScrollBar_TrackColor)';
var sapScrollBar_Hover_FaceColor = 'var(--sapScrollBar_Hover_FaceColor)';
var sapUiFieldReadOnlyHelpBackground = 'var(--sapUiFieldReadOnlyHelpBackground)';
var sapUiButtonRejectActiveBackgroundDarken5 = 'var(--sapUiButtonRejectActiveBackgroundDarken5)';
var sapUiButtonRejectActiveBackgroundLighten5 = 'var(--sapUiButtonRejectActiveBackgroundLighten5)';
var sapUiButtonAcceptActiveBackgroundDarken5 = 'var(--sapUiButtonAcceptActiveBackgroundDarken5)';
var sapUiButtonAcceptActiveBackgroundLighten5 = 'var(--sapUiButtonAcceptActiveBackgroundLighten5)';
var sapField_Hover_HelpBackground = 'var(--sapField_Hover_HelpBackground)';
var sapUiButtonHoverBackground = 'var(--sapUiButtonHoverBackground)';
var sapUiErrorBGLighten4 = 'var(--sapUiErrorBGLighten4)';
var sapUiSuccessBGLighten5 = 'var(--sapUiSuccessBGLighten5)';
var sapUiContentForegroundColorLighten5 = 'var(--sapUiContentForegroundColorLighten5)';
var sapUiContentForegroundColorLighten7 = 'var(--sapUiContentForegroundColorLighten7)';
var sapUiContentForegroundColorDarken3 = 'var(--sapUiContentForegroundColorDarken3)';
var sapUiContentForegroundColorDarken5 = 'var(--sapUiContentForegroundColorDarken5)';
var sapUiContentForegroundColorDarken10 = 'var(--sapUiContentForegroundColorDarken10)';
var sapUiScrollBarTrackColor = 'var(--sapUiScrollBarTrackColor)';
var sapUiScrollBarHoverFaceColor = 'var(--sapUiScrollBarHoverFaceColor)';
var sapUiInfobarHoverBackground = 'var(--sapUiInfobarHoverBackground)';
var sapUiListTableGroupHeaderBorderColor = 'var(--sapUiListTableGroupHeaderBorderColor)';
var sapUiListTableFixedBorder = 'var(--sapUiListTableFixedBorder)';
var sapUiListBorderColorLighten10 = 'var(--sapUiListBorderColorLighten10)';
var sapUiSelectedDarken10 = 'var(--sapUiSelectedDarken10)';
var sapUiActiveLighten3 = 'var(--sapUiActiveLighten3)';
var sapUiFieldHoverHelpBackground = 'var(--sapUiFieldHoverHelpBackground)';
var sapUiButtonLiteHoverBackground = 'var(--sapUiButtonLiteHoverBackground)';
var sapUiSegmentedButtonHoverBackground = 'var(--sapUiSegmentedButtonHoverBackground)';
var sapUiButtonHoverBorderColorLighten30 = 'var(--sapUiButtonHoverBorderColorLighten30)';
var sapUiCalloutShadow = 'var(--sapUiCalloutShadow)';
var sapUiShadowHeader = 'var(--sapUiShadowHeader)';
var sapUiShadowLevel0 = 'var(--sapUiShadowLevel0)';
var sapUiToggleButtonPressedHoverBackground = 'var(--sapUiToggleButtonPressedHoverBackground)';
var sapUiButtonLiteHoverBorderColor = 'var(--sapUiButtonLiteHoverBorderColor)';
var sapUiButtonFooterHoverBackground = 'var(--sapUiButtonFooterHoverBackground)';
var sapUiSegmentedButtonFooterHoverBackground = 'var(--sapUiSegmentedButtonFooterHoverBackground)';
var sapUiToggleButtonPressedHoverBorderColor = 'var(--sapUiToggleButtonPressedHoverBorderColor)';
var sapUiSegmentedButtonSelectedHoverBackground = 'var(--sapUiSegmentedButtonSelectedHoverBackground)';
var sapList_Hover_Background = 'var(--sapList_Hover_Background)';
var sapUiSegmentedButtonSelectedHoverBorderColor = 'var(--sapUiSegmentedButtonSelectedHoverBorderColor)';
var sapList_SelectionBackgroundColor = 'var(--sapList_SelectionBackgroundColor)';
var sapUiListHoverBackground = 'var(--sapUiListHoverBackground)';
var sapUiListSelectionBackgroundColor = 'var(--sapUiListSelectionBackgroundColor)';
var sapUiButtonLiteActionSelectHoverBackground = 'var(--sapUiButtonLiteActionSelectHoverBackground)';
var sapUiToggleButtonPressedBackgroundLighten50Desaturate47 = 'var(--sapUiToggleButtonPressedBackgroundLighten50Desaturate47)';
var sapUiButtonHoverBackgroundDarken2 = 'var(--sapUiButtonHoverBackgroundDarken2)';
var sapUiButtonHoverBackgroundDarken5 = 'var(--sapUiButtonHoverBackgroundDarken5)';
var sapUiToggleButtonPressedBorderColorLighten19Desaturate46 = 'var(--sapUiToggleButtonPressedBorderColorLighten19Desaturate46)';
var sapShell_TextColor = 'var(--sapShell_TextColor)';
var sapButton_Emphasized_TextColor = 'var(--sapButton_Emphasized_TextColor)';
var sapButton_TextColor = 'var(--sapButton_TextColor)';
var sapPageHeader_TextColor = 'var(--sapPageHeader_TextColor)';
var sapPageFooter_TextColor = 'var(--sapPageFooter_TextColor)';
var sapTile_TitleTextColor = 'var(--sapTile_TitleTextColor)';
var sapTile_TextColor = 'var(--sapTile_TextColor)';
var sapTile_IconColor = 'var(--sapTile_IconColor)';
var sapUiShellTextColor = 'var(--sapUiShellTextColor)';
var sapUiShadowLevel1 = 'var(--sapUiShadowLevel1)';
var sapUiShadowLevel2 = 'var(--sapUiShadowLevel2)';
var sapUiShadowLevel3 = 'var(--sapUiShadowLevel3)';
var sapUiButtonEmphasizedTextColor = 'var(--sapUiButtonEmphasizedTextColor)';
var sapUiButtonTextColor = 'var(--sapUiButtonTextColor)';
var sapUiPageHeaderTextColor = 'var(--sapUiPageHeaderTextColor)';
var sapUiPageFooterBorderColor = 'var(--sapUiPageFooterBorderColor)';
var sapUiPageFooterTextColor = 'var(--sapUiPageFooterTextColor)';
var sapUiTileTitleTextColor = 'var(--sapUiTileTitleTextColor)';
var sapUiTileTextColor = 'var(--sapUiTileTextColor)';
var sapUiTileIconColor = 'var(--sapUiTileIconColor)';
var sapUiButtonRejectTextColor = 'var(--sapUiButtonRejectTextColor)';
var sapUiButtonAcceptTextColor = 'var(--sapUiButtonAcceptTextColor)';
var sapUiButtonLiteTextColor = 'var(--sapUiButtonLiteTextColor)';
var sapUiButtonHeaderTextColor = 'var(--sapUiButtonHeaderTextColor)';
var sapUiButtonIconColor = 'var(--sapUiButtonIconColor)';
var sapUiSegmentedButtonIconColor = 'var(--sapUiSegmentedButtonIconColor)';
var sapUiButtonHeaderDisabledTextColor = 'var(--sapUiButtonHeaderDisabledTextColor)';
var sapHighlightTextColor = 'var(--sapHighlightTextColor)';
var sapGroup_TitleTextColor = 'var(--sapGroup_TitleTextColor)';
var sapContent_ForegroundTextColor = 'var(--sapContent_ForegroundTextColor)';
var sapUiShellActiveTextColor = 'var(--sapUiShellActiveTextColor)';
var sapList_HeaderTextColor = 'var(--sapList_HeaderTextColor)';
var sapUiHighlightTextColor = 'var(--sapUiHighlightTextColor)';
var sapUiGroupTitleTextColor = 'var(--sapUiGroupTitleTextColor)';
var sapUiContentForegroundTextColor = 'var(--sapUiContentForegroundTextColor)';
var sapUiListTableGroupHeaderTextColor = 'var(--sapUiListTableGroupHeaderTextColor)';
var sapUiListHeaderTextColor = 'var(--sapUiListHeaderTextColor)';
var sapUiListActiveTextColor = 'var(--sapUiListActiveTextColor)';
var sapUiSegmentedButtonSelectedIconColor = 'var(--sapUiSegmentedButtonSelectedIconColor)';
var sapUiSegmentedButtonActiveIconColor = 'var(--sapUiSegmentedButtonActiveIconColor)';
var sapButton_Hover_TextColor = 'var(--sapButton_Hover_TextColor)';
var sapUiListSelectionHoverBackground = 'var(--sapUiListSelectionHoverBackground)';
var sapUiButtonHoverTextColor = 'var(--sapUiButtonHoverTextColor)';
var sapUiButtonFooterTextColor = 'var(--sapUiButtonFooterTextColor)';
var sapUiSegmentedButtonTextColor = 'var(--sapUiSegmentedButtonTextColor)';
var sapUiToggleButtonPressedTextColor = 'var(--sapUiToggleButtonPressedTextColor)';
var sapUiButtonActiveTextColor = 'var(--sapUiButtonActiveTextColor)';
var sapUiSegmentedButtonSelectedTextColor = 'var(--sapUiSegmentedButtonSelectedTextColor)';
var sapUiSegmentedButtonActiveTextColor = 'var(--sapUiSegmentedButtonActiveTextColor)';

export { sapAccentColor1, sapAccentColor10, sapAccentColor2, sapAccentColor3, sapAccentColor4, sapAccentColor5, sapAccentColor6, sapAccentColor7, sapAccentColor8, sapAccentColor9, sapActiveColor, sapBackgroundColor, sapBackgroundColorDefault, sapBackgroundColorFade72, sapBackgroundImage, sapBackgroundImageOpacity, sapBackgroundImageRepeat, sapBaseColor, sapBlockLayer_Background, sapBrandColor, sapButton_Accept_Background, sapButton_Background, sapButton_BorderColor, sapButton_BorderCornerRadius, sapButton_BorderWidth, sapButton_Emphasized_Background, sapButton_Emphasized_BorderColor, sapButton_Emphasized_TextColor, sapButton_Hover_Background, sapButton_Hover_BorderColor, sapButton_Hover_TextColor, sapButton_Reject_Background, sapButton_TextColor, sapCompanyLogo, sapContent_BadgeBackground, sapContent_ContrastFocusColor, sapContent_ContrastIconColor, sapContent_ContrastShadowColor, sapContent_ContrastTextColor, sapContent_ContrastTextThreshold, sapContent_DisabledOpacity, sapContent_DisabledTextColor, sapContent_ElementHeight, sapContent_ElementHeight_PX, sapContent_FocusColor, sapContent_ForegroundBorderColor, sapContent_ForegroundColor, sapContent_ForegroundTextColor, sapContent_GridSize, sapContent_HelpColor, sapContent_IconColor, sapContent_IconHeight, sapContent_ImagePlaceholderBackground, sapContent_ImagePlaceholderForegroundColor, sapContent_LabelColor, sapContent_LineHeight, sapContent_MarkerIconColor, sapContent_MarkerTextColor, sapContent_MonospaceFontFamily, sapContent_NonInteractiveIconColor, sapContent_SearchHighlightColor, sapContent_ShadowColor, sapCriticalColor, sapCriticalElementColor, sapCriticalTextColor, sapElement_BorderWidth, sapElement_Height, sapElement_LineHeight, sapErrorBackground, sapErrorBorderColor, sapErrorColor, sapField_Background, sapField_BorderColor, sapField_BorderCornerRadius, sapField_BorderWidth, sapField_Focus_Background, sapField_Focus_BorderColor, sapField_Focus_HelpBackground, sapField_HelpBackground, sapField_Hover_Background, sapField_Hover_BorderColor, sapField_Hover_HelpBackground, sapField_InvalidBackground, sapField_InvalidColor, sapField_ReadOnly_Background, sapField_ReadOnly_BorderColor, sapField_ReadOnly_HelpBackground, sapField_RequiredColor, sapField_SuccessBackground, sapField_SuccessColor, sapField_TextColor, sapField_WarningBackground, sapField_WarningColor, sapFontFamily, sapFontSize, sapGroup_BorderCornerRadius, sapGroup_BorderWidth, sapGroup_ContentBackground, sapGroup_ContentBorderColor, sapGroup_FooterBackground, sapGroup_TitleBackground, sapGroup_TitleBorderColor, sapGroup_TitleTextColor, sapGroup_Title_FontSize, sapHighlightColor, sapHighlightTextColor, sapIndicationColor_1, sapIndicationColor_2, sapIndicationColor_3, sapIndicationColor_4, sapIndicationColor_5, sapInfobar_Background, sapInformationBackground, sapInformationBorderColor, sapInformationColor, sapInformativeColor, sapInformativeElementColor, sapInformativeTextColor, sapLinkColor, sapList_Background, sapList_BorderColor, sapList_BorderWidth, sapList_HeaderBackground, sapList_HeaderBorderColor, sapList_HeaderTextColor, sapList_HighlightColor, sapList_Hover_Background, sapList_SelectionBackgroundColor, sapMFontHeader1Size, sapMFontHeader2Size, sapMFontHeader3Size, sapMFontHeader4Size, sapMFontHeader5Size, sapMFontHeader6Size, sapMFontLargeSize, sapMFontMediumSize, sapMFontSmallSize, sapMPlatformDependent, sapNegativeColor, sapNegativeElementColor, sapNegativeTextColor, sapNeutralBackground, sapNeutralBorderColor, sapNeutralColor, sapNeutralElementColor, sapNeutralTextColor, sapObjectHeader_Background, sapPageFooter_Background, sapPageFooter_TextColor, sapPageHeader_Background, sapPageHeader_BorderColor, sapPageHeader_TextColor, sapPositiveColor, sapPositiveElementColor, sapPositiveTextColor, sapPrimary1, sapPrimary2, sapPrimary3, sapPrimary4, sapPrimary5, sapPrimary6, sapPrimary7, sapScrollBar_BorderColor, sapScrollBar_Dimension, sapScrollBar_FaceColor, sapScrollBar_Hover_FaceColor, sapScrollBar_SymbolColor, sapScrollBar_TrackColor, sapSelectedColor, sapShellColor, sapShell_Background, sapShell_BackgroundGradient, sapShell_BackgroundImage, sapShell_BackgroundImageOpacity, sapShell_BackgroundImageRepeat, sapShell_BackgroundPatternColor, sapShell_BorderColor, sapShell_Favicon, sapShell_InteractiveTextColor, sapShell_TextColor, sapSuccessBackground, sapSuccessBorderColor, sapSuccessColor, sapTextColor, sapTile_Background, sapTile_BorderColor, sapTile_IconColor, sapTile_TextColor, sapTile_TitleTextColor, sapTitleColor, sapToolbar_Background, sapToolbar_SeparatorColor, sapUiAccent1, sapUiAccent10, sapUiAccent10Lighten49, sapUiAccent1Lighten50, sapUiAccent2, sapUiAccent2Lighten40, sapUiAccent3, sapUiAccent3Lighten46, sapUiAccent4, sapUiAccent4Lighten46, sapUiAccent5, sapUiAccent5Lighten32, sapUiAccent6, sapUiAccent6Lighten52, sapUiAccent7, sapUiAccent7Lighten64, sapUiAccent8, sapUiAccent8Lighten61, sapUiAccent9, sapUiAccent9Lighten37, sapUiActive, sapUiActiveLighten3, sapUiBackgroundImage, sapUiBaseBG, sapUiBaseColor, sapUiBaseText, sapUiBlockLayerBackground, sapUiBrand, sapUiButtonAcceptActiveBackground, sapUiButtonAcceptActiveBackgroundDarken5, sapUiButtonAcceptActiveBackgroundLighten5, sapUiButtonAcceptActiveBorderColor, sapUiButtonAcceptBackground, sapUiButtonAcceptBorderColor, sapUiButtonAcceptHoverBackground, sapUiButtonAcceptHoverBorderColor, sapUiButtonAcceptTextColor, sapUiButtonActionSelectBackground, sapUiButtonActionSelectBorderColor, sapUiButtonActiveBackground, sapUiButtonActiveBorderColor, sapUiButtonActiveTextColor, sapUiButtonBackground, sapUiButtonBackgroundDarken10, sapUiButtonBackgroundDarken2, sapUiButtonBackgroundDarken24, sapUiButtonBackgroundDarken7, sapUiButtonBorderColor, sapUiButtonBorderCornerRadius, sapUiButtonBorderWidth, sapUiButtonEmphasizedActiveBackground, sapUiButtonEmphasizedActiveBorderColor, sapUiButtonEmphasizedBackground, sapUiButtonEmphasizedBorderColor, sapUiButtonEmphasizedHoverBackground, sapUiButtonEmphasizedHoverBorderColor, sapUiButtonEmphasizedTextColor, sapUiButtonEmphasizedTextShadow, sapUiButtonFooterHoverBackground, sapUiButtonFooterTextColor, sapUiButtonHeaderDisabledTextColor, sapUiButtonHeaderTextColor, sapUiButtonHoverBackground, sapUiButtonHoverBackgroundDarken2, sapUiButtonHoverBackgroundDarken5, sapUiButtonHoverBorderColor, sapUiButtonHoverBorderColorLighten30, sapUiButtonHoverTextColor, sapUiButtonIconColor, sapUiButtonLiteActionSelectHoverBackground, sapUiButtonLiteActiveBackground, sapUiButtonLiteActiveBorderColor, sapUiButtonLiteBackground, sapUiButtonLiteBorderColor, sapUiButtonLiteHoverBackground, sapUiButtonLiteHoverBorderColor, sapUiButtonLiteTextColor, sapUiButtonRejectActiveBackground, sapUiButtonRejectActiveBackgroundDarken5, sapUiButtonRejectActiveBackgroundLighten5, sapUiButtonRejectActiveBorderColor, sapUiButtonRejectBackground, sapUiButtonRejectBorderColor, sapUiButtonRejectHoverBackground, sapUiButtonRejectHoverBorderColor, sapUiButtonRejectTextColor, sapUiButtonTextColor, sapUiCalendarColorToday, sapUiCalloutShadow, sapUiContentBadgeBackground, sapUiContentContrastFocusColor, sapUiContentContrastIconColor, sapUiContentContrastShadowColor, sapUiContentContrastTextColor, sapUiContentContrastTextThreshold, sapUiContentDisabledOpacity, sapUiContentDisabledTextColor, sapUiContentElementHeight, sapUiContentElementHeightPX, sapUiContentFocusColor, sapUiContentForegroundBorderColor, sapUiContentForegroundColor, sapUiContentForegroundColorDarken10, sapUiContentForegroundColorDarken3, sapUiContentForegroundColorDarken5, sapUiContentForegroundColorLighten5, sapUiContentForegroundColorLighten7, sapUiContentForegroundTextColor, sapUiContentGridSize, sapUiContentHelpColor, sapUiContentIconColor, sapUiContentIconHeight, sapUiContentImagePlaceholderBackground, sapUiContentImagePlaceholderForegroundColor, sapUiContentLabelColor, sapUiContentLineHeight, sapUiContentMarkerIconColor, sapUiContentMarkerTextColor, sapUiContentNonInteractiveIconColor, sapUiContentSearchHighlightColor, sapUiContentShadowColor, sapUiContentShadowColorFade15, sapUiContentShadowColorFade30, sapUiContentShadowColorFade5, sapUiCritical, sapUiCriticalElement, sapUiCriticalText, sapUiDesktopFontFamily, sapUiDesktopFontSize, sapUiDragAndDropActiveBackground, sapUiDragAndDropActiveBorderColor, sapUiDragAndDropActiveColor, sapUiDragAndDropBackground, sapUiDragAndDropBorderColor, sapUiDragAndDropColor, sapUiElementBorderWidth, sapUiElementHeight, sapUiElementLineHeight, sapUiErrorBG, sapUiErrorBGLighten4, sapUiErrorBorder, sapUiFieldActiveBackground, sapUiFieldActiveBorderColor, sapUiFieldActiveTextColor, sapUiFieldBackground, sapUiFieldBorderColor, sapUiFieldBorderCornerRadius, sapUiFieldBorderWidth, sapUiFieldFocusBackground, sapUiFieldFocusBorderColor, sapUiFieldFocusHelpBackground, sapUiFieldHelpBackground, sapUiFieldHoverBackground, sapUiFieldHoverBorderColor, sapUiFieldHoverHelpBackground, sapUiFieldInvalidBackground, sapUiFieldInvalidColor, sapUiFieldPlaceholderTextColor, sapUiFieldReadOnlyBackground, sapUiFieldReadOnlyBorderColor, sapUiFieldReadOnlyHelpBackground, sapUiFieldRequiredColor, sapUiFieldSuccessBackground, sapUiFieldSuccessColor, sapUiFieldTextColor, sapUiFieldWarningBackground, sapUiFieldWarningColor, sapUiFieldWarningColorDarken100, sapUiFontCondensedFamily, sapUiFontFamily, sapUiFontHeader1Size, sapUiFontHeader2Size, sapUiFontHeader3Size, sapUiFontHeader4Size, sapUiFontHeader5Size, sapUiFontHeader6Size, sapUiFontHeaderFamily, sapUiFontHeaderWeight, sapUiFontLargeSize, sapUiFontSize, sapUiFontSmallSize, sapUiGlobalBackgroundColor, sapUiGlobalBackgroundImage, sapUiGlobalBackgroundImageOpacity, sapUiGlobalBackgroundRepeat, sapUiGlobalLogo, sapUiGroupBorderWidth, sapUiGroupContentBackground, sapUiGroupContentBorderColor, sapUiGroupFooterBackground, sapUiGroupTitleBackground, sapUiGroupTitleBorderColor, sapUiGroupTitleTextColor, sapUiHighlight, sapUiHighlightTextColor, sapUiIndication1, sapUiIndication2, sapUiIndication3, sapUiIndication4, sapUiIndication5, sapUiInfobarActiveBackground, sapUiInfobarBackground, sapUiInfobarHoverBackground, sapUiInformationBG, sapUiInformationBorder, sapUiInformative, sapUiInformativeElement, sapUiInformativeText, sapUiLineHeight, sapUiLink, sapUiLinkActive, sapUiLinkDarken15, sapUiLinkHover, sapUiLinkInverted, sapUiLinkVisited, sapUiListActiveBackground, sapUiListActiveTextColor, sapUiListBackground, sapUiListBackgroundDarken10, sapUiListBackgroundDarken13, sapUiListBackgroundDarken15, sapUiListBackgroundDarken20, sapUiListBackgroundDarken3, sapUiListBorderColor, sapUiListBorderColorLighten10, sapUiListBorderWidth, sapUiListFooterBackground, sapUiListFooterTextColor, sapUiListGroupHeaderBackground, sapUiListHeaderBackground, sapUiListHeaderBorderColor, sapUiListHeaderTextColor, sapUiListHighlightColor, sapUiListHoverBackground, sapUiListSelectionBackgroundColor, sapUiListSelectionHoverBackground, sapUiListTableFixedBorder, sapUiListTableFooterBorder, sapUiListTableGroupHeaderBackground, sapUiListTableGroupHeaderBorderColor, sapUiListTableGroupHeaderTextColor, sapUiListTableIconSize, sapUiListTableTextSize, sapUiListTextColor, sapUiListVerticalBorderColor, sapUiNegative, sapUiNegativeElement, sapUiNegativeText, sapUiNeutral, sapUiNeutralBG, sapUiNeutralBorder, sapUiNeutralElement, sapUiNeutralText, sapUiNotificationBarBG, sapUiNotificationBarBorder, sapUiNotifierSeparator, sapUiNotifierSeparatorWidth, sapUiObjectHeaderBackground, sapUiObjectHeaderBorderColor, sapUiPageFooterBackground, sapUiPageFooterBorderColor, sapUiPageFooterTextColor, sapUiPageHeaderBackground, sapUiPageHeaderBorderColor, sapUiPageHeaderTextColor, sapUiPositive, sapUiPositiveElement, sapUiPositiveText, sapUiPrimary1, sapUiPrimary2, sapUiPrimary3, sapUiPrimary4, sapUiPrimary5, sapUiPrimary6, sapUiPrimary7, sapUiScrollBarBorderColor, sapUiScrollBarDimension, sapUiScrollBarFaceColor, sapUiScrollBarHoverFaceColor, sapUiScrollBarSymbolColor, sapUiScrollBarTrackColor, sapUiSegmentedButtonActiveBackground, sapUiSegmentedButtonActiveIconColor, sapUiSegmentedButtonActiveTextColor, sapUiSegmentedButtonBackground, sapUiSegmentedButtonBorderColor, sapUiSegmentedButtonFooterBorderColor, sapUiSegmentedButtonFooterHoverBackground, sapUiSegmentedButtonHoverBackground, sapUiSegmentedButtonIconColor, sapUiSegmentedButtonSelectedBackground, sapUiSegmentedButtonSelectedHoverBackground, sapUiSegmentedButtonSelectedHoverBorderColor, sapUiSegmentedButtonSelectedIconColor, sapUiSegmentedButtonSelectedTextColor, sapUiSegmentedButtonTextColor, sapUiSelected, sapUiSelectedDarken10, sapUiShadowHeader, sapUiShadowLevel0, sapUiShadowLevel1, sapUiShadowLevel2, sapUiShadowLevel3, sapUiShadowText, sapUiShellActiveBackground, sapUiShellActiveTextColor, sapUiShellAltContainerBackground, sapUiShellBackground, sapUiShellBackgroundGradient, sapUiShellBackgroundImage, sapUiShellBackgroundImageOpacity, sapUiShellBackgroundImageRepeat, sapUiShellBackgroundPatternColor, sapUiShellBorderColor, sapUiShellBorderColorLighten30, sapUiShellColor, sapUiShellContainerBackground, sapUiShellFavicon, sapUiShellGroupTextColor, sapUiShellHoverBackground, sapUiShellHoverToggleBackground, sapUiShellInteractiveTextColor, sapUiShellTextColor, sapUiSuccessBG, sapUiSuccessBGLighten5, sapUiSuccessBorder, sapUiTextTitle, sapUiTileBackground, sapUiTileBackgroundDarken20, sapUiTileBorderColor, sapUiTileIconColor, sapUiTileTextColor, sapUiTileTitleTextColor, sapUiToggleButtonPressedBackground, sapUiToggleButtonPressedBackgroundLighten50Desaturate47, sapUiToggleButtonPressedBorderColor, sapUiToggleButtonPressedBorderColorLighten19Desaturate46, sapUiToggleButtonPressedHoverBackground, sapUiToggleButtonPressedHoverBorderColor, sapUiToggleButtonPressedTextColor, sapUiToolbarBackground, sapUiToolbarSeparatorColor, sapUiTranslucentBGOpacity, sapUiUx3ShellBackgroundImageURL, sapUiWarningBG, sapUiWarningBorder, sapWarningBackground, sapWarningBorderColor, sapWarningColor, sap_wc_input_compact_min_width, sap_wc_input_disabled_opacity, sap_wc_input_icon_min_width };
